 <!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>List of algorithms available in pygmo &#8212; pagmo 2.8 documentation</title>
    <link rel="stylesheet" href="../../../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pagmo.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '2.8',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../../../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../../../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../../../_static/bootstrap-3.3.7/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../../../_static/bootstrap-sphinx.js"></script>
    <link rel="shortcut icon" href="../../../_static/logo_favico.ico"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="List of islands available in pygmo" href="../islands/py_islands.html" />
    <link rel="prev" title="List of problems available in pygmo" href="../problems/py_problems.html" />
<link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Roboto|Comfortaa|Space+Mono"> 
  </head>
  <body>

  <div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../../../index.html"><span><img src="../../../_static/logo_small_html.png"></span>
           </a>
        <span class="navbar-text navbar-version pull-left"><b>2.8</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../../../index.html">Home</a></li>
                <li><a href="../../../install.html">Install</a></li>
                <li><a href="../../../quickstart.html">Quickstart</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../../../index.html">All <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul>
<li class="toctree-l1"><a class="reference internal" href="../../../install.html">Installation guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#c">C++</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../install.html#python">Python</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../quickstart.html">Quick start</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../quickstart.html#c">C++</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../quickstart.html#python">Python</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../algorithm_list.html">List of algorithms (UDAs) available in pagmo/pygmo</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../algorithm_list.html#heuristic-global-optimization">Heuristic Global Optimization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../algorithm_list.html#meta-algorithms">Meta-algorithms</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../algorithm_list.html#local-optimization">Local optimization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../problem_list.html">List of problems (UDPs) available in pagmo/pygmo</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../problem_list.html#scalable-problems">Scalable problems</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../problem_list.html#problem-suites">Problem suites</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../problem_list.html#meta-problems">Meta-problems</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../island_list.html">List of islands (UDIs) available in pagmo/pygmo</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../cpp/cpp_docs.html">C++ Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../cpp/cpp_docs.html#core-classes">Core classes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../cpp/cpp_docs.html#implemented-algorithms">Implemented algorithms</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../cpp/cpp_docs.html#implemented-problems">Implemented problems</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../cpp/cpp_docs.html#implemented-islands">Implemented islands</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../cpp/cpp_docs.html#utilities">Utilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../cpp/cpp_docs.html#miscellanea">Miscellanea</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../python_docs.html">Python Documentation</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../python_docs.html#core-classes">Core classes</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../python_docs.html#implemented-problems-algorithms-and-islands">Implemented problems, algorithms and islands</a></li>
<li class="toctree-l2"><a class="reference internal" href="../python_docs.html#utilities">Utilities</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/python_tut.html">Python tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/python_tut.html#basics">Basics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/python_tut.html#coding-your-own-problem-udp">Coding your own problem (UDP)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/python_tut.html#coding-your-own-island-udi">Coding your own island (UDI)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/python_tut.html#local-optimization">Local optimization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/python_tut.html#multi-objective-optimization">Multi-objective optimization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/python_tut.html#hypervolumes">Hypervolumes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/python_tut.html#advanced-examples">Advanced Examples</a></li>
</ul>
</li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../credits.html">Credits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../credits.html#main-developers">Main developers</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../changelog.html#unreleased">2.8 (unreleased)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../changelog.html#id1">2.7 (2018-04-13)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../changelog.html#id4">2.6 (2017-11-18)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../changelog.html#id6">2.5 (2017-11-12)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../changelog.html#id8">2.4 (2017-06-09)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../changelog.html#id12">2.3 (2017-05-19)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../changelog.html#id15">2.2 (2017-05-12)</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">This Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">List of algorithms available in pygmo</a><ul>
<li><a class="reference internal" href="#a-quick-look">A Quick Look</a><ul>
<li><a class="reference internal" href="#heuristic-global-optimization">Heuristic Global Optimization</a></li>
<li><a class="reference internal" href="#meta-algorithms">Meta-algorithms</a></li>
<li><a class="reference internal" href="#local-optimization">Local optimization</a></li>
</ul>
</li>
<li><a class="reference internal" href="#algorithms-exposed-from-c">Algorithms exposed from C++</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../../../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="list-of-algorithms-available-in-pygmo">
<span id="py-algorithms"></span><h1>List of algorithms available in pygmo<a class="headerlink" href="#list-of-algorithms-available-in-pygmo" title="Permalink to this headline">¶</a></h1>
<div class="section" id="a-quick-look">
<h2>A Quick Look<a class="headerlink" href="#a-quick-look" title="Permalink to this headline">¶</a></h2>
<p>User defined algorithms (UDAs) are classes that allow the construction of a <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">pygmo.algorithm</span></code></a> offering a unified access to
all of them. The user can implement his own UDA, or use any ot the ones we provide in pygmo and that are listed here for convenience.</p>
<p>Each algorithm can be associated only to problems of certain types S = Single, M = Multi-objective, C = Constrained, U = Unconstrained, sto = stochastic</p>
<div class="section" id="heuristic-global-optimization">
<h3>Heuristic Global Optimization<a class="headerlink" href="#heuristic-global-optimization" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="32%" />
<col width="23%" />
<col width="8%" />
<col width="37%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Common Name</th>
<th class="head">Name in PyGMO</th>
<th class="head">Type</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>The null algorithm</td>
<td><a class="reference internal" href="#pygmo.null_algorithm" title="pygmo.null_algorithm"><code class="xref py py-class docutils literal"><span class="pre">pygmo.null_algorithm</span></code></a></td>
<td>SM-CU</td>
<td>Exposed from C++, used for initialization purposes, does nothing</td>
</tr>
<tr class="row-odd"><td>Differential Evolution (DE)</td>
<td><a class="reference internal" href="#pygmo.de" title="pygmo.de"><code class="xref py py-class docutils literal"><span class="pre">pygmo.de</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>Self-adaptive DE (jDE and iDE)</td>
<td><a class="reference internal" href="#pygmo.sade" title="pygmo.sade"><code class="xref py py-class docutils literal"><span class="pre">pygmo.sade</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>Self-adaptive DE (de_1220 aka pDE)</td>
<td><a class="reference internal" href="#pygmo.de1220" title="pygmo.de1220"><code class="xref py py-class docutils literal"><span class="pre">pygmo.de1220</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>Particle Swarm Optimization (PSO)</td>
<td><a class="reference internal" href="#pygmo.pso" title="pygmo.pso"><code class="xref py py-class docutils literal"><span class="pre">pygmo.pso</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>Particle Swarm Optimization Generational (GPSO)</td>
<td><a class="reference internal" href="#pygmo.pso_gen" title="pygmo.pso_gen"><code class="xref py py-class docutils literal"><span class="pre">pygmo.pso_gen</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>(N+1)-ES Simple Evolutionary Algorithm</td>
<td><a class="reference internal" href="#pygmo.sea" title="pygmo.sea"><code class="xref py py-class docutils literal"><span class="pre">pygmo.sea</span></code></a></td>
<td>S-U (sto)</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>Simple Genetic Algorithm</td>
<td><a class="reference internal" href="#pygmo.sga" title="pygmo.sga"><code class="xref py py-class docutils literal"><span class="pre">pygmo.sga</span></code></a></td>
<td>S-U (sto)</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>Corana’s Simulated Annealing (SA)</td>
<td><a class="reference internal" href="#pygmo.simulated_annealing" title="pygmo.simulated_annealing"><code class="xref py py-class docutils literal"><span class="pre">pygmo.simulated_annealing</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>Artificial Bee Colony (ABC)</td>
<td><a class="reference internal" href="#pygmo.bee_colony" title="pygmo.bee_colony"><code class="xref py py-class docutils literal"><span class="pre">pygmo.bee_colony</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>Covariance Matrix Adaptation Evo. Strategy (CMA-ES)</td>
<td><a class="reference internal" href="#pygmo.cmaes" title="pygmo.cmaes"><code class="xref py py-class docutils literal"><span class="pre">pygmo.cmaes</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>Non-dominated Sorting GA (NSGA2)</td>
<td><a class="reference internal" href="#pygmo.nsga2" title="pygmo.nsga2"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nsga2</span></code></a></td>
<td>M-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>Multi-objective EA vith Decomposition (MOEA/D)</td>
<td><a class="reference internal" href="#pygmo.moead" title="pygmo.moead"><code class="xref py py-class docutils literal"><span class="pre">pygmo.moead</span></code></a></td>
<td>M-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>Improved Harmony Serach</td>
<td><a class="reference internal" href="#pygmo.ihs" title="pygmo.ihs"><code class="xref py py-class docutils literal"><span class="pre">pygmo.ihs</span></code></a></td>
<td>SM-CU-I</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>Exponential Evolution Strategies (xNES)</td>
<td><a class="reference internal" href="#pygmo.xnes" title="pygmo.xnes"><code class="xref py py-class docutils literal"><span class="pre">pygmo.xnes</span></code></a></td>
<td>S-U (sto)</td>
<td>Exposed from C++</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="meta-algorithms">
<h3>Meta-algorithms<a class="headerlink" href="#meta-algorithms" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="25%" />
<col width="32%" />
<col width="11%" />
<col width="32%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Common Name</th>
<th class="head">Name in PyGMO</th>
<th class="head">Type</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Monotonic Basin Hopping (MBH)</td>
<td><a class="reference internal" href="#pygmo.mbh" title="pygmo.mbh"><code class="xref py py-class docutils literal"><span class="pre">pygmo.mbh</span></code></a></td>
<td>S-CU</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>Cstrs Self-Adaptive</td>
<td><a class="reference internal" href="#pygmo.cstrs_self_adaptive" title="pygmo.cstrs_self_adaptive"><code class="xref py py-class docutils literal"><span class="pre">pygmo.cstrs_self_adaptive</span></code></a></td>
<td>S-C</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>Augmented Lagrangian algorithm</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-CU</td>
<td>Exposed from C++</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="local-optimization">
<h3>Local optimization<a class="headerlink" href="#local-optimization" title="Permalink to this headline">¶</a></h3>
<table border="1" class="docutils">
<colgroup>
<col width="30%" />
<col width="23%" />
<col width="8%" />
<col width="39%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Common Name</th>
<th class="head">Name in PyGMO</th>
<th class="head">Type</th>
<th class="head">Comments</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Compass Search (CS)</td>
<td><a class="reference internal" href="#pygmo.compass_search" title="pygmo.compass_search"><code class="xref py py-class docutils literal"><span class="pre">pygmo.compass_search</span></code></a></td>
<td>S-CU</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>COBYLA (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-CU</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>BOBYQA (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>NEWUOA + bound constraints (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>PRAXIS (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>Nelder-Mead simplex (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>sbplx (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>MMA (Method of Moving Asymptotes) (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-CU</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>CCSA (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-CU</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>SLSQP (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-CU</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>low-storage BFGS (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>preconditioned truncated Newton (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-even"><td>Shifted limited-memory variable-metric (from NLopt)</td>
<td><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.nlopt</span></code></a></td>
<td>S-U</td>
<td>Exposed from C++</td>
</tr>
<tr class="row-odd"><td>Ipopt</td>
<td><a class="reference internal" href="#pygmo.ipopt" title="pygmo.ipopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.ipopt</span></code></a></td>
<td>S-CU</td>
<td>Exposed from C++</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr class="docutils" />
<div class="section" id="algorithms-exposed-from-c">
<h2>Algorithms exposed from C++<a class="headerlink" href="#algorithms-exposed-from-c" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="pygmo.null_algorithm">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">null_algorithm</code><a class="headerlink" href="#pygmo.null_algorithm" title="Permalink to this definition">¶</a></dt>
<dd><p>The null algorithm.</p>
<p>An algorithm used in the default-initialization of <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">pygmo.algorithm</span></code></a> and of the meta-algorithms.</p>
</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.bee_colony">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">bee_colony</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>limit = 1</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.bee_colony" title="Permalink to this definition">¶</a></dt>
<dd><p>Artificial Bee Colony.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>limit</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – maximum number of trials for abandoning a source</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em>, <em>limit</em> or <em>seed</em> is negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>limit</em> is not greater than 0</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/bee_colony.html#_CPPv2N5pagmo10bee_colonyE" title="pagmo::bee_colony"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::bee_colony</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.bee_colony.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.bee_colony.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code>. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.bee_colony" title="pygmo.bee_colony"><code class="xref py py-class docutils literal"><span class="pre">bee_colony</span></code></a>. A verbosity of <code class="docutils literal"><span class="pre">N</span></code> implies a log line each <code class="docutils literal"><span class="pre">N</span></code> generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Current</span> <span class="pre">best</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">Current</span> <span class="pre">best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness currently in the population</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness found</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">bee_colony</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span> <span class="n">limit</span> <span class="o">=</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">rosenbrock</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:          Best: Current Best:</span>
<span class="go">   1             40         261363         261363</span>
<span class="go"> 101           4040        112.237        267.969</span>
<span class="go"> 201           8040        20.8885        265.122</span>
<span class="go"> 301          12040        20.6076        20.6076</span>
<span class="go"> 401          16040         18.252        140.079</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">bee_colony</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 40, 183727.83934515435, 183727.83934515435), ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/bee_colony.html#_CPPv2NK5pagmo10bee_colony7get_logEv" title="pagmo::bee_colony::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::bee_colony::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.bee_colony.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.bee_colony.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.de">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">de</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>F = 0.8</em>, <em>CR = 0.9</em>, <em>variant = 2</em>, <em>ftol = 1e-6</em>, <em>xtol = 1e-6</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.de" title="Permalink to this definition">¶</a></dt>
<dd><p>Differential Evolution</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>F</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – weight coefficient (dafault value is 0.8)</li>
<li><strong>CR</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – crossover probability (dafault value is 0.9)</li>
<li><strong>variant</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – mutation variant (dafault variant is 2: /rand/1/exp)</li>
<li><strong>ftol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the f tolerance (default is 1e-6)</li>
<li><strong>xtol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the x tolerance (default is 1e-6)</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em>, <em>variant</em> or <em>seed</em> is negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>F</em>, <em>CR</em> are not in [0,1] or <em>variant</em> is not in [1, 10]</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>The following variants (mutation variants) are available to create a new candidate individual:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1 - best/1/exp</td>
<td>2 - rand/1/exp</td>
</tr>
<tr class="row-even"><td>3 - rand-to-best/1/exp</td>
<td>4 - best/2/exp</td>
</tr>
<tr class="row-odd"><td>5 - rand/2/exp</td>
<td>6 - best/1/bin</td>
</tr>
<tr class="row-even"><td>7 - rand/1/bin</td>
<td>8 - rand-to-best/1/bin</td>
</tr>
<tr class="row-odd"><td>9 - best/2/bin</td>
<td>10 - rand/2/bin</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/de.html#_CPPv2N5pagmo2deE" title="pagmo::de"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::de</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.de.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.de.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code>. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.de" title="pygmo.de"><code class="xref py py-class docutils literal"><span class="pre">de</span></code></a>. A verbosity of <code class="docutils literal"><span class="pre">N</span></code> implies a log line each <code class="docutils literal"><span class="pre">N</span></code> generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">dx</span></code>, <code class="docutils literal"><span class="pre">df</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function currently in the population</li>
<li><code class="docutils literal"><span class="pre">dx</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the norm of the distance to the population mean of the mutant vectors</li>
<li><code class="docutils literal"><span class="pre">df</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the population flatness evaluated as the distance between the fitness of the best and of the worst individual</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">de</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">rosenbrock</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:          Best:            dx:            df:</span>
<span class="go">  1             20         162446        65.2891    1.78686e+06</span>
<span class="go">101           2020        198.402         8.4454        572.161</span>
<span class="go">201           4020        21.1155        2.60629        24.5152</span>
<span class="go">301           6020        6.67069        0.51811        1.99744</span>
<span class="go">401           8020        3.60022       0.583444       0.554511</span>
<span class="go">Exit condition -- generations = 500</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">de</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 20, 162446.0185265718, 65.28911664703388, 1786857.8926660626), ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/de.html#_CPPv2NK5pagmo2de7get_logEv" title="pagmo::de::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::de::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.de.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.de.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.sea">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">sea</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.sea" title="Permalink to this definition">¶</a></dt>
<dd><p>(N+1)-ES simple evolutionary algorithm.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations to consider (each generation will compute the objective function once)</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em> or <em>seed</em> are negative or greater than an implementation-defined value</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python
(e.g., type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/sea.html#_CPPv2N5pagmo3seaE" title="pagmo::sea"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::sea</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.sea.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.sea.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code> and printed to screen.
The log frequency depends on the verbosity parameter (by default nothing is logged) which can be set calling
the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> constructed with a
<a class="reference internal" href="#pygmo.sea" title="pygmo.sea"><code class="xref py py-class docutils literal"><span class="pre">sea</span></code></a>.
A verbosity larger than 1 will produce a log with one entry each verbosity fitness evaluations.
A verbosity equal to 1 will produce a log with one entry at each improvement of the fitness.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">Improvement</span></code>, <code class="docutils literal"><span class="pre">Mutations</span></code><ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation.</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made.</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function found so far.</li>
<li><code class="docutils literal"><span class="pre">Improvement</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), improvement made by the last mutation.</li>
<li><code class="docutils literal"><span class="pre">Mutations</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), number of mutated components for the decision vector.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">sea</span><span class="p">(</span><span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">schwefel</span><span class="p">(</span><span class="n">dim</span> <span class="o">=</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:          Best:   Improvement:     Mutations:</span>
<span class="go">   1              1        6363.44        2890.49              2</span>
<span class="go">1001           1001        1039.92       -562.407              3</span>
<span class="go">2001           2001        358.966         -632.6              2</span>
<span class="go">3001           3001         106.08       -995.927              3</span>
<span class="go">4001           4001         83.391         -266.8              1</span>
<span class="go">5001           5001        62.4994       -1018.38              3</span>
<span class="go">6001           6001        39.2851       -732.695              2</span>
<span class="go">7001           7001        37.2185       -518.847              1</span>
<span class="go">8001           8001        20.9452        -450.75              1</span>
<span class="go">9001           9001        17.9193       -270.679              1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">sea</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 1, 6363.442036625835, 2890.4854414320716, 2), (1001, 1001, ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/sea.html#_CPPv2NK5pagmo3sea7get_logEv" title="pagmo::sea::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::sea::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.sea.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.sea.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.sga">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">sga</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>cr = .90</em>, <em>eta_c = 1.</em>, <em>m = 0.02</em>, <em>param_m = 1.</em>, <em>param_s = 2</em>, <em>crossover = &quot;exponential&quot;</em>, <em>mutation = &quot;polynomial&quot;</em>, <em>selection = &quot;tournament&quot;</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.sga" title="Permalink to this definition">¶</a></dt>
<dd><p>A Simple Genetic Algorithm</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.2.</span></p>
</div>
<p>Approximately during the same decades as Evolutionary Strategies (see <a class="reference internal" href="#pygmo.sea" title="pygmo.sea"><code class="xref py py-class docutils literal"><span class="pre">sea</span></code></a>) were studied,
a different group led by John Holland, and later by his student David Goldberg, introduced and
studied an algorithmic framework called “genetic algorithms” that were, essentially, leveraging on
the same idea but introducing also crossover as a genetic operator. This led to a few decades of
confusion and discussions on what was an evolutionary startegy and what a genetic algorithm and on
whether the crossover was a useful operator or mutation only algorithms were to be preferred.</p>
<p>In pygmo we provide a rather classical implementation of a genetic algorithm, letting the user choose between
selected crossover types, selection schemes and mutation types.</p>
<p>The pseudo code of our version is:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>&gt; Start from a population (pop) of dimension N
&gt; while i &lt; gen
&gt; &gt; Selection: create a new population (pop2) with N individuals selected from pop (with repetition allowed)
&gt; &gt; Crossover: create a new population (pop3) with N individuals obtained applying crossover to pop2
&gt; &gt; Mutation:  create a new population (pop4) with N individuals obtained applying mutation to pop3
&gt; &gt; Evaluate all new chromosomes in pop4
&gt; &gt; Reinsertion: set pop to contain the best N individuals taken from pop and pop4
</pre></div>
</div>
<p>The various blocks of pygmo genetic algorithm are listed below:</p>
<p><em>Selection</em>: two selection methods are provided: <code class="docutils literal"><span class="pre">tournament</span></code> and <code class="docutils literal"><span class="pre">truncated</span></code>. <code class="docutils literal"><span class="pre">Tournament</span></code> selection works by
selecting each offspring as the one having the minimal fitness in a random group of size <em>param_s</em>. The <code class="docutils literal"><span class="pre">truncated</span></code>
selection, instead, works selecting the best <em>param_s</em> chromosomes in the entire population over and over.
We have deliberately not implemented the popular roulette wheel selection as we are of the opinion that such
a system does not generalize much being highly sensitive to the fitness scaling.</p>
<p><em>Crossover</em>: four different crossover schemes are provided:<code class="docutils literal"><span class="pre">single</span></code>, <code class="docutils literal"><span class="pre">exponential</span></code>, <code class="docutils literal"><span class="pre">binomial</span></code>, <code class="docutils literal"><span class="pre">sbx</span></code>. The
<code class="docutils literal"><span class="pre">single</span></code> point crossover, works selecting a random point in the parent chromosome and, with probability <em>cr</em>, inserting the
partner chromosome thereafter. The <code class="docutils literal"><span class="pre">exponential</span></code> crossover is taken from the algorithm differential evolution,
implemented, in pygmo, as <a class="reference internal" href="#pygmo.de" title="pygmo.de"><code class="xref py py-class docutils literal"><span class="pre">de</span></code></a>. It essentially selects a random point in the parent chromosome and inserts,
in each successive gene, the partner values with probability  <em>cr</em> up to when it stops. The binomial crossover
inserts each gene from the partner with probability <em>cr</em>. The simulated binary crossover (called <code class="docutils literal"><span class="pre">sbx</span></code>), is taken
from the NSGA-II algorithm, implemented in pygmo as <a class="reference internal" href="#pygmo.nsga2" title="pygmo.nsga2"><code class="xref py py-class docutils literal"><span class="pre">nsga2</span></code></a>, and makes use of an additional parameter called
distribution index <em>eta_c</em>.</p>
<p><em>Mutation</em>: three different mutations schemes are provided: <code class="docutils literal"><span class="pre">uniform</span></code>, <code class="docutils literal"><span class="pre">gaussian</span></code> and <code class="docutils literal"><span class="pre">polynomial</span></code>. Uniform mutation
simply randomly samples from the bounds. Gaussian muattion samples around each gene using a normal distribution
with standard deviation proportional to the <em>param_m</em> and the bounds width. The last scheme is the <code class="docutils literal"><span class="pre">polynomial</span></code>
mutation from Deb.</p>
<p><em>Reinsertion</em>: the only reinsertion strategy provided is what we call pure elitism. After each generation
all parents and children are put in the same pool and only the best are passed to the next generation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations.</li>
<li><strong>cr</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – crossover probability.</li>
<li><strong>eta_c</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – distribution index for <code class="docutils literal"><span class="pre">sbx</span></code> crossover. This parameter is inactive if other types of crossover are selected.</li>
<li><strong>m</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – mutation probability.</li>
<li><strong>param_m</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – distribution index (<code class="docutils literal"><span class="pre">polynomial</span></code> mutation), gaussian width (<code class="docutils literal"><span class="pre">gaussian</span></code> mutation) or inactive (<code class="docutils literal"><span class="pre">uniform</span></code> mutation)</li>
<li><strong>param_s</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – the number of best individuals to use in “truncated” selection or the size of the tournament in <code class="docutils literal"><span class="pre">tournament</span></code> selection.</li>
<li><strong>crossover</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – the crossover strategy. One of <code class="docutils literal"><span class="pre">exponential</span></code>, <code class="docutils literal"><span class="pre">binomial</span></code>, <code class="docutils literal"><span class="pre">single</span></code> or <code class="docutils literal"><span class="pre">sbx</span></code></li>
<li><strong>mutation</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – the mutation strategy. One of <code class="docutils literal"><span class="pre">gaussian</span></code>, <code class="docutils literal"><span class="pre">polynomial</span></code> or <code class="docutils literal"><span class="pre">uniform</span></code>.</li>
<li><strong>selection</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – the selection strategy. One of <code class="docutils literal"><span class="pre">tournament</span></code>, “truncated”.</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em> or <em>seed</em> are negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>cr</em> is not in [0,1], if <em>eta_c</em> is not in [1,100], if <em>m</em> is not in [0,1], input_f <em>mutation</em>
is not one of <code class="docutils literal"><span class="pre">gaussian</span></code>, <code class="docutils literal"><span class="pre">uniform</span></code> or <code class="docutils literal"><span class="pre">polynomial</span></code>, if <em>selection</em> not one of “roulette”,
“truncated” or <em>crossover</em> is not one of <code class="docutils literal"><span class="pre">exponential</span></code>, <code class="docutils literal"><span class="pre">binomial</span></code>, <code class="docutils literal"><span class="pre">sbx</span></code>, <code class="docutils literal"><span class="pre">single</span></code>, if <em>param_m</em> is
not in [0,1] and <em>mutation</em> is not <code class="docutils literal"><span class="pre">polynomial</span></code>, if <em>mutation</em> is not in [1,100] and <em>mutation</em> is <code class="docutils literal"><span class="pre">polynomial</span></code>.</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python
(e.g., type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/sga.html#_CPPv2N5pagmo3sgaE" title="pagmo::sga"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::sga</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.sga.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.sga.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code> and printed to screen.
The log frequency depends on the verbosity parameter (by default nothing is logged) which can be set calling
the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> constructed with a
<a class="reference internal" href="#pygmo.sga" title="pygmo.sga"><code class="xref py py-class docutils literal"><span class="pre">sga</span></code></a>.
A verbosity larger than 1 will produce a log with one entry each verbosity fitness evaluations.
A verbosity equal to 1 will produce a log with one entry at each improvement of the fitness.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">Improvement</span></code><p><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation.
<code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made.
<code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function found so far.
<code class="docutils literal"><span class="pre">Improvement</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), improvement made by the last generation.</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">sga</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">schwefel</span><span class="p">(</span><span class="n">dim</span> <span class="o">=</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:          Best:   Improvement:     Mutations:</span>
<span class="go">   1              1        6363.44        2890.49              2</span>
<span class="go">1001           1001        1039.92       -562.407              3</span>
<span class="go">2001           2001        358.966         -632.6              2</span>
<span class="go">3001           3001         106.08       -995.927              3</span>
<span class="go">4001           4001         83.391         -266.8              1</span>
<span class="go">5001           5001        62.4994       -1018.38              3</span>
<span class="go">6001           6001        39.2851       -732.695              2</span>
<span class="go">7001           7001        37.2185       -518.847              1</span>
<span class="go">8001           8001        20.9452        -450.75              1</span>
<span class="go">9001           9001        17.9193       -270.679              1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">sea</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 1, 6363.442036625835, 2890.4854414320716, 2), (1001, 1001, ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/sga.html#_CPPv2NK5pagmo3sga7get_logEv" title="pagmo::sga::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::sga::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.sga.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.sga.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.sade">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">sade</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>variant = 2</em>, <em>variant_adptv = 1</em>, <em>ftol = 1e-6</em>, <em>xtol = 1e-6</em>, <em>memory = False</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.sade" title="Permalink to this definition">¶</a></dt>
<dd><p>Self-adaptive Differential Evolution.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>variant</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – mutation variant (dafault variant is 2: /rand/1/exp)</li>
<li><strong>variant_adptv</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – F and CR parameter adaptation scheme to be used (one of 1..2)</li>
<li><strong>ftol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the x tolerance (default is 1e-6)</li>
<li><strong>xtol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the f tolerance (default is 1e-6)</li>
<li><strong>memory</strong> (<code class="docutils literal"><span class="pre">bool</span></code>) – when true the adapted parameters CR anf F are not reset between successive calls to the evolve method</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em>, <em>variant</em>, <em>variant_adptv</em> or <em>seed</em> is negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>variant</em> is not in [1,18] or <em>variant_adptv</em> is not in [0,1]</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>The following variants (mutation variants) are available to create a new candidate individual:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1 - best/1/exp</td>
<td>2 - rand/1/exp</td>
</tr>
<tr class="row-even"><td>3 - rand-to-best/1/exp</td>
<td>4 - best/2/exp</td>
</tr>
<tr class="row-odd"><td>5 - rand/2/exp</td>
<td>6 - best/1/bin</td>
</tr>
<tr class="row-even"><td>7 - rand/1/bin</td>
<td>8 - rand-to-best/1/bin</td>
</tr>
<tr class="row-odd"><td>9 - best/2/bin</td>
<td>10 - rand/2/bin</td>
</tr>
<tr class="row-even"><td>11 - rand/3/exp</td>
<td>12 - rand/3/bin</td>
</tr>
<tr class="row-odd"><td>13 - best/3/exp</td>
<td>14 - best/3/bin</td>
</tr>
<tr class="row-even"><td>15 - rand-to-current/2/exp</td>
<td>16 - rand-to-current/2/bin</td>
</tr>
<tr class="row-odd"><td>17 - rand-to-best-and-current/2/exp</td>
<td>18 - rand-to-best-and-current/2/bin</td>
</tr>
</tbody>
</table>
<p>The following adaptation schemes are available:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1 - jDE</td>
<td>2 - iDE</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/sade.html#_CPPv2N5pagmo4sadeE" title="pagmo::sade"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::sade</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.sade.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.sade.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code>. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.sade" title="pygmo.sade"><code class="xref py py-class docutils literal"><span class="pre">sade</span></code></a>. A verbosity of <code class="docutils literal"><span class="pre">N</span></code> implies a log line each <code class="docutils literal"><span class="pre">N</span></code> generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">F</span></code>, <code class="docutils literal"><span class="pre">CR</span></code>, <code class="docutils literal"><span class="pre">dx</span></code>, <code class="docutils literal"><span class="pre">df</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function currently in the population</li>
<li><code class="docutils literal"><span class="pre">F</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the value of the adapted paramter F used to create the best so far</li>
<li><code class="docutils literal"><span class="pre">CR</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the value of the adapted paramter CR used to create the best so far</li>
<li><code class="docutils literal"><span class="pre">dx</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the norm of the distance to the population mean of the mutant vectors</li>
<li><code class="docutils literal"><span class="pre">df</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the population flatness evaluated as the distance between the fitness of the best and of the worst individual</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">sade</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problems</span><span class="o">.</span><span class="n">rosenbrock</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:          Best:             F:            CR:            dx:            df:</span>
<span class="go">  1             20         297060       0.690031       0.294769        44.1494    2.30584e+06</span>
<span class="go">101           2020        97.4258        0.58354       0.591527        13.3115        441.545</span>
<span class="go">201           4020        8.79247         0.6678        0.53148        17.8822        121.676</span>
<span class="go">301           6020        6.84774       0.494549        0.98105        12.2781        40.9626</span>
<span class="go">401           8020         4.7861       0.428741       0.743813        12.2938        39.7791</span>
<span class="go">Exit condition -- generations = 500</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">sade</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 20, 297059.6296130389, 0.690031071850855, 0.29476914701127666, 44.14940516578547, 2305836.7422693395), ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/sade.html#_CPPv2NK5pagmo4sade7get_logEv" title="pagmo::sade::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::sade::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.sade.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.sade.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.de1220">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">de1220</code><span class="sig-paren">(</span><em>gen = 1, allowed_variants = [2,3,7,10,13,14,15,16], variant_adptv = 1, ftol = 1e-6, xtol = 1e-6, memory = False, seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.de1220" title="Permalink to this definition">¶</a></dt>
<dd><p>Self-adaptive Differential Evolution, pygmo flavour (pDE).
The adaptation of the mutation variant is added to <a class="reference internal" href="#pygmo.sade" title="pygmo.sade"><code class="xref py py-class docutils literal"><span class="pre">sade</span></code></a></p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>allowed_variants</strong> (<em>array-like object</em>) – allowed mutation variants, each one being a number in [1, 18]</li>
<li><strong>variant_adptv</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – <em>F</em> and <em>CR</em> parameter adaptation scheme to be used (one of 1..2)</li>
<li><strong>ftol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the x tolerance (default is 1e-6)</li>
<li><strong>xtol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the f tolerance (default is 1e-6)</li>
<li><strong>memory</strong> (<code class="docutils literal"><span class="pre">bool</span></code>) – when true the adapted parameters <em>CR</em> anf <em>F</em> are not reset between successive calls to the evolve method</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em>, <em>variant</em>, <em>variant_adptv</em> or <em>seed</em> is negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if each id in <em>variant_adptv</em> is not in [1,18] or <em>variant_adptv</em> is not in [0,1]</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>The following variants (mutation variants) can be put into <em>allowed_variants</em>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1 - best/1/exp</td>
<td>2 - rand/1/exp</td>
</tr>
<tr class="row-even"><td>3 - rand-to-best/1/exp</td>
<td>4 - best/2/exp</td>
</tr>
<tr class="row-odd"><td>5 - rand/2/exp</td>
<td>6 - best/1/bin</td>
</tr>
<tr class="row-even"><td>7 - rand/1/bin</td>
<td>8 - rand-to-best/1/bin</td>
</tr>
<tr class="row-odd"><td>9 - best/2/bin</td>
<td>10 - rand/2/bin</td>
</tr>
<tr class="row-even"><td>11 - rand/3/exp</td>
<td>12 - rand/3/bin</td>
</tr>
<tr class="row-odd"><td>13 - best/3/exp</td>
<td>14 - best/3/bin</td>
</tr>
<tr class="row-even"><td>15 - rand-to-current/2/exp</td>
<td>16 - rand-to-current/2/bin</td>
</tr>
<tr class="row-odd"><td>17 - rand-to-best-and-current/2/exp</td>
<td>18 - rand-to-best-and-current/2/bin</td>
</tr>
</tbody>
</table>
<p>The following adaptation schemes for the parameters <em>F</em> and <em>CR</em> are available:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1 - jDE</td>
<td>2 - iDE</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/de1220.html#_CPPv2N5pagmo6de1220E" title="pagmo::de1220"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::de1220</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.de1220.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.de1220.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code>. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.de1220" title="pygmo.de1220"><code class="xref py py-class docutils literal"><span class="pre">de1220</span></code></a>. A verbosity of N implies a log line each N generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">F</span></code>, <code class="docutils literal"><span class="pre">CR</span></code>, <code class="docutils literal"><span class="pre">Variant</span></code>, <code class="docutils literal"><span class="pre">dx</span></code>, <code class="docutils literal"><span class="pre">df</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function currently in the population</li>
<li><code class="docutils literal"><span class="pre">F</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the value of the adapted paramter F used to create the best so far</li>
<li><code class="docutils literal"><span class="pre">CR</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the value of the adapted paramter CR used to create the best so far</li>
<li><code class="docutils literal"><span class="pre">Variant</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), the mutation variant used to create the best so far</li>
<li><code class="docutils literal"><span class="pre">dx</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the norm of the distance to the population mean of the mutant vectors</li>
<li><code class="docutils literal"><span class="pre">df</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the population flatness evaluated as the distance between the fitness of the best and of the worst individual</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">de1220</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">rosenbrock</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:          Best:             F:            CR:       Variant:            dx:            df:</span>
<span class="go">    1             20         285653        0.55135       0.441551             16        43.9719    2.02379e+06</span>
<span class="go">101           2020        12.2721       0.127285      0.0792493             14        3.22986        106.764</span>
<span class="go">201           4020        5.72927       0.148337       0.777806             14        2.72177        4.10793</span>
<span class="go">301           6020        4.85084        0.12193       0.996191              3        2.95555        3.85027</span>
<span class="go">401           8020        4.20638       0.235997       0.996259              3        3.60338        4.49432</span>
<span class="go">Exit condition -- generations = 500</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">de1220</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 20, 285652.7928977573, 0.551350234239449, 0.4415510963067054, 16, 43.97185788345982, 2023791.5123259544), ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/de1220.html#_CPPv2NK5pagmo6de12207get_logEv" title="pagmo::de1220::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::de1220::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.de1220.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.de1220.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.cmaes">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">cmaes</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>cc = -1</em>, <em>cs = -1</em>, <em>c1 = -1</em>, <em>cmu = -1</em>, <em>sigma0 = 0.5</em>, <em>ftol = 1e-6</em>, <em>xtol = 1e-6</em>, <em>memory = False</em>, <em>force_bounds = False</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.cmaes" title="Permalink to this definition">¶</a></dt>
<dd><p>Covariance Matrix Evolutionary Strategy (CMA-ES).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>cc</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – backward time horizon for the evolution path (by default is automatically assigned)</li>
<li><strong>cs</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – makes partly up for the small variance loss in case the indicator is zero (by default is automatically assigned)</li>
<li><strong>c1</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – learning rate for the rank-one update of the covariance matrix (by default is automatically assigned)</li>
<li><strong>cmu</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – learning rate for the rank-mu  update of the covariance matrix (by default is automatically assigned)</li>
<li><strong>sigma0</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – initial step-size</li>
<li><strong>ftol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the x tolerance</li>
<li><strong>xtol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the f tolerance</li>
<li><strong>memory</strong> (<code class="docutils literal"><span class="pre">bool</span></code>) – when true the adapted parameters are not reset between successive calls to the evolve method</li>
<li><strong>force_bounds</strong> (<code class="docutils literal"><span class="pre">bool</span></code>) – when true the box bounds are enforced. The fitness will never be called outside the bounds but the covariance matrix adaptation  mechanism will worsen</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em> is negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>cc</em>, <em>cs</em>, <em>c1</em>, <em>cmu</em> are not in [0,1] or -1</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/cmaes.html#_CPPv2N5pagmo5cmaesE" title="pagmo::cmaes"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::cmaes</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.cmaes.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.cmaes.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code>. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.cmaes" title="pygmo.cmaes"><code class="xref py py-class docutils literal"><span class="pre">cmaes</span></code></a>. A verbosity of <code class="docutils literal"><span class="pre">N</span></code> implies a log line each <code class="docutils literal"><span class="pre">N</span></code> generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">dx</span></code>, <code class="docutils literal"><span class="pre">df</span></code>, <code class="docutils literal"><span class="pre">sigma</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function currently in the population</li>
<li><code class="docutils literal"><span class="pre">dx</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the norm of the distance to the population mean of the mutant vectors</li>
<li><code class="docutils literal"><span class="pre">df</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the population flatness evaluated as the distance between the fitness of the best and of the worst individual</li>
<li><code class="docutils literal"><span class="pre">sigma</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the current step-size</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">cmaes</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">rosenbrock</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:          Best:            dx:            df:         sigma:</span>
<span class="go">  1              0         173924        33.6872    3.06519e+06            0.5</span>
<span class="go">101           2000        92.9612       0.583942        156.921      0.0382078</span>
<span class="go">201           4000        8.79819       0.117574          5.101      0.0228353</span>
<span class="go">301           6000        4.81377      0.0698366        1.34637      0.0297664</span>
<span class="go">401           8000        1.04445      0.0568541       0.514459      0.0649836</span>
<span class="go">Exit condition -- generations = 500</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">cmaes</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 0, 173924.2840042722, 33.68717961390855, 3065192.3843070837, 0.5), ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/cmaes.html#_CPPv2NK5pagmo5cmaes7get_logEv" title="pagmo::cmaes::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::cmaes::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.cmaes.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.cmaes.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.moead">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">moead</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>weight_generation = &quot;grid&quot;</em>, <em>decomposition = &quot;tchebycheff&quot;</em>, <em>neighbours = 20</em>, <em>CR = 1</em>, <em>F = 0.5</em>, <em>eta_m = 20</em>, <em>realb = 0.9</em>, <em>limit = 2</em>, <em>preserve_diversity = true</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.moead" title="Permalink to this definition">¶</a></dt>
<dd><p>Multi Objective Evolutionary Algorithms by Decomposition (the DE variant)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>weight_generation</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – method used to generate the weights, one of “grid”, “low discrepancy” or “random”</li>
<li><strong>decomposition</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – method used to decompose the objectives, one of “tchebycheff”, “weighted” or “bi”</li>
<li><strong>neighbours</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – size of the weight’s neighborhood</li>
<li><strong>CR</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – crossover parameter in the Differential Evolution operator</li>
<li><strong>F</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – parameter for the Differential Evolution operator</li>
<li><strong>eta_m</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – distribution index used by the polynomial mutation</li>
<li><strong>realb</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – chance that the neighbourhood is considered at each generation, rather than the whole population (only if preserve_diversity is true)</li>
<li><strong>limit</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – maximum number of copies reinserted in the population  (only if m_preserve_diversity is true)</li>
<li><strong>preserve_diversity</strong> (<code class="docutils literal"><span class="pre">bool</span></code>) – when true activates diversity preservation mechanisms</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em>, <em>neighbours</em>, <em>seed</em> or <em>limit</em> are negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if either:<ul>
<li><em>decomposition</em> is not one of ‘tchebycheff’, ‘weighted’ or ‘bi’.</li>
<li><em>weight_generation</em> is not one of ‘random’, ‘low discrepancy’ or ‘grid’.</li>
<li><em>CR</em> or <em>F</em> or <em>realb</em> are not in [0.,1.]</li>
<li><em>eta_m</em> is negative</li>
</ul>
</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/moead.html#_CPPv2N5pagmo5moeadE" title="pagmo::moead"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::moead</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.moead.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.moead.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code>. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.moead" title="pygmo.moead"><code class="xref py py-class docutils literal"><span class="pre">moead</span></code></a>. A verbosity of <code class="docutils literal"><span class="pre">N</span></code> implies a log line each <code class="docutils literal"><span class="pre">N</span></code> generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">ADR</span></code>, <code class="docutils literal"><span class="pre">ideal_point</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">ADF</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), Average Decomposed Fitness, that is the average across all decomposed problem of the single objective decomposed fitness along the corresponding direction</li>
<li><code class="docutils literal"><span class="pre">ideal_point</span></code> (<code class="docutils literal"><span class="pre">array</span></code>), the ideal point of the current population (cropped to max 5 dimensions only in the screen output)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">moead</span><span class="p">(</span><span class="n">gen</span><span class="o">=</span><span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">zdt</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">40</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:           ADF:        ideal1:        ideal2:</span>
<span class="go">  1              0        32.5747     0.00190532        2.65685</span>
<span class="go">101           4000        5.67751    2.56736e-09       0.468789</span>
<span class="go">201           8000        5.38297    2.56736e-09      0.0855025</span>
<span class="go">301          12000        5.05509    9.76581e-10      0.0574796</span>
<span class="go">401          16000        5.13126    9.76581e-10      0.0242256</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">moead</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 0, 32.574745630075874, array([  1.90532430e-03,   2.65684834e+00])), ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/moead.html#_CPPv2NK5pagmo5moead7get_logEv" title="pagmo::moead::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::moead::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.moead.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.moead.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.compass_search">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">compass_search</code><span class="sig-paren">(</span><em>max_fevals = 1</em>, <em>start_range = .1</em>, <em>stop_range = .01</em>, <em>reduction_coeff = .5</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.compass_search" title="Permalink to this definition">¶</a></dt>
<dd><p>Compass Search</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>max_fevals</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – maximum number of function evaluation</li>
<li><strong>start_range</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – start range (dafault value is .1)</li>
<li><strong>stop_range</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stop range (dafault value is .01)</li>
<li><strong>reduction_coeff</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – range reduction coefficient (dafault value is .5)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>max_fevals</em> is negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>start_range</em> is not in (0, 1], if <em>stop_range</em> is not in (<em>start_range</em>, 1] or if <em>reduction_coeff</em> is not in (0,1)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/compass_search.html#_CPPv2N5pagmo14compass_searchE" title="pagmo::compass_search"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::compass_search</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.compass_search.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.compass_search.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code> and printed to screen. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.compass_search" title="pygmo.compass_search"><code class="xref py py-class docutils literal"><span class="pre">compass_search</span></code></a>. A verbosity larger than 0 implies one log line at each improvment of the fitness or
change in the search range.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values``Fevals``, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">Range</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function currently in the population</li>
<li><code class="docutils literal"><span class="pre">Range</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the range used to vary the chromosome (relative to the box bounds width)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">compass_search</span><span class="p">(</span><span class="n">max_fevals</span> <span class="o">=</span> <span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">hock_schittkowsky_71</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Fevals:          Best:      Violated:    Viol. Norm:         Range:</span>
<span class="go">      4        110.785              1        2.40583            0.5</span>
<span class="go">     12        110.785              1        2.40583           0.25</span>
<span class="go">     20        110.785              1        2.40583          0.125</span>
<span class="go">     22        91.0454              1        1.01855          0.125</span>
<span class="go">     25        96.2795              1       0.229446          0.125</span>
<span class="go">     33        96.2795              1       0.229446         0.0625</span>
<span class="go">     41        96.2795              1       0.229446        0.03125</span>
<span class="go">     45         94.971              1       0.127929        0.03125</span>
<span class="go">     53         94.971              1       0.127929       0.015625</span>
<span class="go">     56        95.6252              1      0.0458521       0.015625</span>
<span class="go">     64        95.6252              1      0.0458521      0.0078125</span>
<span class="go">     68        95.2981              1      0.0410151      0.0078125</span>
<span class="go">     76        95.2981              1      0.0410151     0.00390625</span>
<span class="go">     79        95.4617              1     0.00117433     0.00390625</span>
<span class="go">     87        95.4617              1     0.00117433     0.00195312</span>
<span class="go">     95        95.4617              1     0.00117433    0.000976562</span>
<span class="go">    103        95.4617              1     0.00117433    0.000488281</span>
<span class="go">    111        95.4617              1     0.00117433    0.000244141</span>
<span class="go">    115        95.4515              0              0    0.000244141</span>
<span class="go">    123        95.4515              0              0     0.00012207</span>
<span class="go">    131        95.4515              0              0    6.10352e-05</span>
<span class="go">    139        95.4515              0              0    3.05176e-05</span>
<span class="go">    143        95.4502              0              0    3.05176e-05</span>
<span class="go">    151        95.4502              0              0    1.52588e-05</span>
<span class="go">    159        95.4502              0              0    7.62939e-06</span>
<span class="go">Exit condition -- range: 7.62939e-06 &lt;= 1e-05</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">compass_search</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(4, 110.785345345, 1, 2.405833534534, 0.5), (12, 110.785345345, 1, 2.405833534534, 0.25) ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/compass_search.html#_CPPv2NK5pagmo14compass_search7get_logEv" title="pagmo::compass_search::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::compass_search::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.compass_search.replacement">
<code class="descname">replacement</code><a class="headerlink" href="#pygmo.compass_search.replacement" title="Permalink to this definition">¶</a></dt>
<dd><p>Individual replacement policy.</p>
<p>This attribute represents the policy that is used in the <code class="docutils literal"><span class="pre">evolve()</span></code> method to select the individual
that will be replaced by the optimised individual. The attribute can be either a string or an integral.</p>
<p>If the attribute is a string, it must be one of <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> will select the best individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> will select the worst individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> will randomly choose one individual in the population.</li>
</ul>
<p><a class="reference internal" href="#pygmo.compass_search.set_random_sr_seed" title="pygmo.compass_search.set_random_sr_seed"><code class="xref py py-func docutils literal"><span class="pre">set_random_sr_seed()</span></code></a> can be used to seed the random number generator
used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> policy.</p>
<p>If the attribute is an integer, it represents the index (in the population) of the individual that will be
replaced by the optimised individual.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the individual replacement policy or index</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">str</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if the attribute is set to an invalid string</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code></a> – if the attribute is set to a value of an invalid type</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.compass_search.selection">
<code class="descname">selection</code><a class="headerlink" href="#pygmo.compass_search.selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Individual selection policy.</p>
<p>This attribute represents the policy that is used in the <code class="docutils literal"><span class="pre">evolve()</span></code> method to select the individual
that will be optimised. The attribute can be either a string or an integral.</p>
<p>If the attribute is a string, it must be one of <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> will select the best individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> will select the worst individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> will randomly choose one individual in the population.</li>
</ul>
<p><a class="reference internal" href="#pygmo.compass_search.set_random_sr_seed" title="pygmo.compass_search.set_random_sr_seed"><code class="xref py py-func docutils literal"><span class="pre">set_random_sr_seed()</span></code></a> can be used to seed the random number generator
used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> policy.</p>
<p>If the attribute is an integer, it represents the index (in the population) of the individual that is selected
for optimisation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the individual selection policy or index</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">str</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if the attribute is set to an invalid string</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code></a> – if the attribute is set to a value of an invalid type</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.compass_search.set_random_sr_seed">
<code class="descname">set_random_sr_seed</code><span class="sig-paren">(</span><em>seed</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.compass_search.set_random_sr_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the seed for the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> selection/replacement policies.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – the value that will be used to seed the random number generator used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>
election/replacement policies (see <a class="reference internal" href="#pygmo.compass_search.selection" title="pygmo.compass_search.selection"><code class="xref py py-attr docutils literal"><span class="pre">selection</span></code></a> and
<a class="reference internal" href="#pygmo.compass_search.replacement" title="pygmo.compass_search.replacement"><code class="xref py py-attr docutils literal"><span class="pre">replacement</span></code></a>)</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.simulated_annealing">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">simulated_annealing</code><span class="sig-paren">(</span><em>Ts = 10.</em>, <em>Tf = .1</em>, <em>n_T_adj = 10</em>, <em>n_range_adj = 10</em>, <em>bin_size = 10</em>, <em>start_range = 1.</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.simulated_annealing" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulated Annealing (Corana’s version)</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>Ts</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – starting temperature</li>
<li><strong>Tf</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – final temperature</li>
<li><strong>n_T_adj</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of temperature adjustments in the annealing schedule</li>
<li><strong>n_range_adj</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of adjustments of the search range performed at a constant temperature</li>
<li><strong>bin_size</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of mutations that are used to compute the acceptance rate</li>
<li><strong>start_range</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – starting range for mutating the decision vector</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>n_T_adj</em>, <em>n_range_adj</em> or <em>bin_size</em> are negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>Ts</em> is not in (0, inf), if <em>Tf</em> is not in (0, inf), if <em>Tf</em> &gt; <em>Ts</em> or if <em>start_range</em> is not in (0,1]</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>n_T_adj</em> is not strictly positive or if <em>n_range_adj</em> is not strictly positive</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/simulated_annealing.html#_CPPv2N5pagmo19simulated_annealingE" title="pagmo::simulated_annealing"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::simulated_annealing</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.simulated_annealing.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.simulated_annealing.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code> and printed to screen.
The log frequency depends on the verbosity parameter (by default nothing is logged) which can be set calling
the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> constructed with a
<a class="reference internal" href="#pygmo.simulated_annealing" title="pygmo.simulated_annealing"><code class="xref py py-class docutils literal"><span class="pre">simulated_annealing</span></code></a>. A verbosity larger than 0 will produce a log with one entry
each verbosity fitness evaluations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">Current</span></code>, <code class="docutils literal"><span class="pre">Mean</span> <span class="pre">range</span></code>, <code class="docutils literal"><span class="pre">Temperature</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function found so far</li>
<li><code class="docutils literal"><span class="pre">Current</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), last fitness sampled</li>
<li><code class="docutils literal"><span class="pre">Mean</span> <span class="pre">range</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the mean search range across the decision vector components (relative to the box bounds width)</li>
<li><code class="docutils literal"><span class="pre">Temperature</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the current temperature</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">simulated_annealing</span><span class="p">(</span><span class="n">Ts</span><span class="o">=</span><span class="mf">10.</span><span class="p">,</span> <span class="n">Tf</span><span class="o">=</span><span class="mf">1e-5</span><span class="p">,</span> <span class="n">n_T_adj</span> <span class="o">=</span> <span class="mi">100</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">5000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">rosenbrock</span><span class="p">(</span><span class="n">dim</span> <span class="o">=</span> <span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Fevals:          Best:       Current:    Mean range:   Temperature:</span>
<span class="go">     57           5937           5937           0.48             10</span>
<span class="go">  10033        9.50937        28.6775      0.0325519        2.51189</span>
<span class="go">  15033        7.87389        14.3951      0.0131132        1.25893</span>
<span class="go">  20033        7.87389        8.68616      0.0120491       0.630957</span>
<span class="go">  25033        2.90084        4.43344     0.00676893       0.316228</span>
<span class="go">  30033       0.963616        1.36471     0.00355931       0.158489</span>
<span class="go">  35033       0.265868        0.63457     0.00202753      0.0794328</span>
<span class="go">  40033        0.13894       0.383283     0.00172611      0.0398107</span>
<span class="go">  45033       0.108051       0.169876    0.000870499      0.0199526</span>
<span class="go">  50033      0.0391731      0.0895308     0.00084195           0.01</span>
<span class="go">  55033      0.0217027      0.0303561    0.000596116     0.00501187</span>
<span class="go">  60033     0.00670073     0.00914824    0.000342754     0.00251189</span>
<span class="go">  65033      0.0012298     0.00791511    0.000275182     0.00125893</span>
<span class="go">  70033     0.00112816     0.00396297    0.000192117    0.000630957</span>
<span class="go">  75033    0.000183055     0.00139717    0.000135137    0.000316228</span>
<span class="go">  80033    0.000174868     0.00192479    0.000109781    0.000158489</span>
<span class="go">  85033       7.83e-05    0.000494225    8.20723e-05    7.94328e-05</span>
<span class="go">  90033    5.35153e-05    0.000120148    5.76009e-05    3.98107e-05</span>
<span class="go">  95033    5.35153e-05    9.10958e-05    3.18624e-05    1.99526e-05</span>
<span class="go">  99933    2.34849e-05    8.72206e-05    2.59215e-05    1.14815e-05</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">simulated_annealing</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(57, 5936.999957947842, 5936.999957947842, 0.47999999999999987, 10.0), (10033, ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/simulated_annealing.html#_CPPv2NK5pagmo19simulated_annealing7get_logEv" title="pagmo::simulated_annealing::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::simulated_annealing::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.simulated_annealing.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.simulated_annealing.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.simulated_annealing.replacement">
<code class="descname">replacement</code><a class="headerlink" href="#pygmo.simulated_annealing.replacement" title="Permalink to this definition">¶</a></dt>
<dd><p>Individual replacement policy.</p>
<p>This attribute represents the policy that is used in the <code class="docutils literal"><span class="pre">evolve()</span></code> method to select the individual
that will be replaced by the optimised individual. The attribute can be either a string or an integral.</p>
<p>If the attribute is a string, it must be one of <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> will select the best individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> will select the worst individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> will randomly choose one individual in the population.</li>
</ul>
<p><a class="reference internal" href="#pygmo.simulated_annealing.set_random_sr_seed" title="pygmo.simulated_annealing.set_random_sr_seed"><code class="xref py py-func docutils literal"><span class="pre">set_random_sr_seed()</span></code></a> can be used to seed the random number generator
used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> policy.</p>
<p>If the attribute is an integer, it represents the index (in the population) of the individual that will be
replaced by the optimised individual.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the individual replacement policy or index</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">str</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if the attribute is set to an invalid string</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code></a> – if the attribute is set to a value of an invalid type</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.simulated_annealing.selection">
<code class="descname">selection</code><a class="headerlink" href="#pygmo.simulated_annealing.selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Individual selection policy.</p>
<p>This attribute represents the policy that is used in the <code class="docutils literal"><span class="pre">evolve()</span></code> method to select the individual
that will be optimised. The attribute can be either a string or an integral.</p>
<p>If the attribute is a string, it must be one of <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> will select the best individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> will select the worst individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> will randomly choose one individual in the population.</li>
</ul>
<p><a class="reference internal" href="#pygmo.simulated_annealing.set_random_sr_seed" title="pygmo.simulated_annealing.set_random_sr_seed"><code class="xref py py-func docutils literal"><span class="pre">set_random_sr_seed()</span></code></a> can be used to seed the random number generator
used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> policy.</p>
<p>If the attribute is an integer, it represents the index (in the population) of the individual that is selected
for optimisation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the individual selection policy or index</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">str</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if the attribute is set to an invalid string</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code></a> – if the attribute is set to a value of an invalid type</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.simulated_annealing.set_random_sr_seed">
<code class="descname">set_random_sr_seed</code><span class="sig-paren">(</span><em>seed</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.simulated_annealing.set_random_sr_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the seed for the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> selection/replacement policies.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – the value that will be used to seed the random number generator used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>
election/replacement policies (see <a class="reference internal" href="#pygmo.simulated_annealing.selection" title="pygmo.simulated_annealing.selection"><code class="xref py py-attr docutils literal"><span class="pre">selection</span></code></a> and
<a class="reference internal" href="#pygmo.simulated_annealing.replacement" title="pygmo.simulated_annealing.replacement"><code class="xref py py-attr docutils literal"><span class="pre">replacement</span></code></a>)</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.pso">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">pso</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>omega = 0.7298</em>, <em>eta1 = 2.05</em>, <em>eta2 = 2.05</em>, <em>max_vel = 0.5</em>, <em>variant = 5</em>, <em>neighb_type = 2</em>, <em>neighb_param = 4</em>, <em>memory = False</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.pso" title="Permalink to this definition">¶</a></dt>
<dd><p>Particle Swarm Optimization</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>omega</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – inertia weight (or constriction factor)</li>
<li><strong>eta1</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – social component</li>
<li><strong>eta2</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – cognitive component</li>
<li><strong>max_vel</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – maximum allowed particle velocities (normalized with respect to the bounds width)</li>
<li><strong>variant</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – algoritmic variant</li>
<li><strong>neighb_type</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – swarm topology (defining each particle’s neighbours)</li>
<li><strong>neighb_param</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – topology parameter (defines how many neighbours to consider)</li>
<li><strong>memory</strong> (<code class="docutils literal"><span class="pre">bool</span></code>) – when true the velocities are not reset between successive calls to the evolve method</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em> or <em>seed</em> is negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>omega</em> is not in the [0,1] interval, if <em>eta1</em>, <em>eta2</em> are not in the [0,1] interval, if <em>max_vel</em> is not in ]0,1]</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – <em>variant</em> is not one of 1 .. 6, if <em>neighb_type</em> is not one of 1 .. 4 or if <em>neighb_param</em> is zero</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>The following variants can be selected via the <em>variant</em> parameter:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1 - Canonical (with inertia weight)</td>
<td>2 - Same social and cognitive rand.</td>
</tr>
<tr class="row-even"><td>3 - Same rand. for all components</td>
<td>4 - Only one rand.</td>
</tr>
<tr class="row-odd"><td>5 - Canonical (with constriction fact.)</td>
<td>6 - Fully Informed (FIPS)</td>
</tr>
</tbody>
</table>
<p>The following topologies are selected by <em>neighb_type</em>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1 - gbest</td>
<td>2 - lbest</td>
</tr>
<tr class="row-even"><td>3 - Von Neumann</td>
<td>4 - Adaptive random</td>
</tr>
</tbody>
</table>
<p>The topology determines (together with the topology parameter) which particles need to be considered
when computing the social component of the velocity update.</p>
<dl class="method">
<dt id="pygmo.pso.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.pso.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code> and printed to screen. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.pso" title="pygmo.pso"><code class="xref py py-class docutils literal"><span class="pre">pso</span></code></a>. A verbosity of <code class="docutils literal"><span class="pre">N</span></code> implies a log line each <code class="docutils literal"><span class="pre">N</span></code> generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">gbest</span></code>, <code class="docutils literal"><span class="pre">Mean</span> <span class="pre">Vel.</span></code>, <code class="docutils literal"><span class="pre">Mean</span> <span class="pre">lbest</span></code>, <code class="docutils literal"><span class="pre">Avg.</span> <span class="pre">Dist.</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">gbest</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function found so far by the the swarm</li>
<li><code class="docutils literal"><span class="pre">Mean</span> <span class="pre">Vel.</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the average particle velocity (normalized)</li>
<li><code class="docutils literal"><span class="pre">Mean</span> <span class="pre">lbest</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the average fitness of the current particle locations</li>
<li><code class="docutils literal"><span class="pre">Avg.</span> <span class="pre">Dist.</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the average distance between particles (normalized)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">pso</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">rosenbrock</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:         gbest:     Mean Vel.:    Mean lbest:    Avg. Dist.:</span>
<span class="go">   1             40        72473.3       0.173892         677427       0.281744</span>
<span class="go">  51           1040        135.867      0.0183806        748.001       0.065826</span>
<span class="go"> 101           2040        12.6726     0.00291046        84.9531      0.0339452</span>
<span class="go"> 151           3040         8.4405    0.000852588        33.5161      0.0191379</span>
<span class="go"> 201           4040        7.56943    0.000778264         28.213     0.00789202</span>
<span class="go"> 251           5040         6.8089     0.00435521        22.7988     0.00107112</span>
<span class="go"> 301           6040         6.3692    0.000289725        17.3763     0.00325571</span>
<span class="go"> 351           7040        6.09414    0.000187343        16.8875     0.00172307</span>
<span class="go"> 401           8040        5.78415    0.000524536        16.5073     0.00234197</span>
<span class="go"> 451           9040         5.4662     0.00018305        16.2339    0.000958182</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">pso</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1,40,72473.32713790605,0.1738915144248373,677427.3504996448,0.2817443174278134), (51,1040,...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/pso.html#_CPPv2NK5pagmo3pso7get_logEv" title="pagmo::pso::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::pso::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.pso.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.pso.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.pso_gen">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">pso_gen</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>omega = 0.7298</em>, <em>eta1 = 2.05</em>, <em>eta2 = 2.05</em>, <em>max_vel = 0.5</em>, <em>variant = 5</em>, <em>neighb_type = 2</em>, <em>neighb_param = 4</em>, <em>memory = False</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.pso_gen" title="Permalink to this definition">¶</a></dt>
<dd><p>Particle Swarm Optimization (generational) is identical to <a class="reference internal" href="#pygmo.pso" title="pygmo.pso"><code class="xref py py-class docutils literal"><span class="pre">pso</span></code></a>, but does update the velocities of each particle before new particle positions are computed (taking
into consideration all updated particle velocities). Each particle is thus evaluated on the same seed within a generation as opposed to the standard PSO which evaluates single particle
at a time. Consequently, the generational PSO algorithm is suited for stochastic optimization problems.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>omega</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – inertia weight (or constriction factor)</li>
<li><strong>eta1</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – social component</li>
<li><strong>eta2</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – cognitive component</li>
<li><strong>max_vel</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – maximum allowed particle velocities (normalized with respect to the bounds width)</li>
<li><strong>variant</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – algoritmic variant</li>
<li><strong>neighb_type</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – swarm topology (defining each particle’s neighbours)</li>
<li><strong>neighb_param</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – topology parameter (defines how many neighbours to consider)</li>
<li><strong>memory</strong> (<code class="docutils literal"><span class="pre">bool</span></code>) – when true the velocities are not reset between successive calls to the evolve method</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em> or <em>seed</em> is negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>omega</em> is not in the [0,1] interval, if <em>eta1</em>, <em>eta2</em> are not in the [0,1] interval, if <em>max_vel</em> is not in ]0,1]</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – <em>variant</em> is not one of 1 .. 6, if <em>neighb_type</em> is not one of 1 .. 4 or if <em>neighb_param</em> is zero</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>The following variants can be selected via the <em>variant</em> parameter:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1 - Canonical (with inertia weight)</td>
<td>2 - Same social and cognitive rand.</td>
</tr>
<tr class="row-even"><td>3 - Same rand. for all components</td>
<td>4 - Only one rand.</td>
</tr>
<tr class="row-odd"><td>5 - Canonical (with constriction fact.)</td>
<td>6 - Fully Informed (FIPS)</td>
</tr>
</tbody>
</table>
<p>The following topologies are selected by <em>neighb_type</em>:</p>
<table border="1" class="docutils">
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>1 - gbest</td>
<td>2 - lbest</td>
</tr>
<tr class="row-even"><td>3 - Von Neumann</td>
<td>4 - Adaptive random</td>
</tr>
</tbody>
</table>
<p>The topology determines (together with the topology parameter) which particles need to be considered
when computing the social component of the velocity update.</p>
<dl class="method">
<dt id="pygmo.pso_gen.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.pso_gen.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code> and printed to screen. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.pso" title="pygmo.pso"><code class="xref py py-class docutils literal"><span class="pre">pso</span></code></a>. A verbosity of <code class="docutils literal"><span class="pre">N</span></code> implies a log line each <code class="docutils literal"><span class="pre">N</span></code> generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">gbest</span></code>, <code class="docutils literal"><span class="pre">Mean</span> <span class="pre">Vel.</span></code>, <code class="docutils literal"><span class="pre">Mean</span> <span class="pre">lbest</span></code>, <code class="docutils literal"><span class="pre">Avg.</span> <span class="pre">Dist.</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">gbest</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function found so far by the the swarm</li>
<li><code class="docutils literal"><span class="pre">Mean</span> <span class="pre">Vel.</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the average particle velocity (normalized)</li>
<li><code class="docutils literal"><span class="pre">Mean</span> <span class="pre">lbest</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the average fitness of the current particle locations</li>
<li><code class="docutils literal"><span class="pre">Avg.</span> <span class="pre">Dist.</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the average distance between particles (normalized)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">pso</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">50</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">rosenbrock</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:         gbest:     Mean Vel.:    Mean lbest:    Avg. Dist.:</span>
<span class="go">   1             40        72473.3       0.173892         677427       0.281744</span>
<span class="go">  51           1040        135.867      0.0183806        748.001       0.065826</span>
<span class="go"> 101           2040        12.6726     0.00291046        84.9531      0.0339452</span>
<span class="go"> 151           3040         8.4405    0.000852588        33.5161      0.0191379</span>
<span class="go"> 201           4040        7.56943    0.000778264         28.213     0.00789202</span>
<span class="go"> 251           5040         6.8089     0.00435521        22.7988     0.00107112</span>
<span class="go"> 301           6040         6.3692    0.000289725        17.3763     0.00325571</span>
<span class="go"> 351           7040        6.09414    0.000187343        16.8875     0.00172307</span>
<span class="go"> 401           8040        5.78415    0.000524536        16.5073     0.00234197</span>
<span class="go"> 451           9040         5.4662     0.00018305        16.2339    0.000958182</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">pso</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1,40,72473.32713790605,0.1738915144248373,677427.3504996448,0.2817443174278134), (51,1040,...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/pso.html#_CPPv2NK5pagmo3pso7get_logEv" title="pagmo::pso::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::pso::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.pso_gen.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.pso_gen.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.nsga2">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">nsga2</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>cr = 0.95</em>, <em>eta_c = 10</em>, <em>m = 0.01</em>, <em>eta_m = 10</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.nsga2" title="Permalink to this definition">¶</a></dt>
<dd><p>Non dominated Sorting Genetic Algorithm (NSGA-II).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>cr</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – crossover probability</li>
<li><strong>eta_c</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – distribution index for crossover</li>
<li><strong>m</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – mutation probability</li>
<li><strong>eta_m</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – distribution index for mutation</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em> or <em>seed</em> are negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if either:<ul>
<li><em>cr</em> is not in [0,1[.</li>
<li><em>eta_c</em> is not in [0,100[.</li>
<li><em>m</em> is not in [0,1].</li>
<li><em>eta_m</em> is not in [0,100[.</li>
</ul>
</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/nsga2.html#_CPPv2N5pagmo5nsga2E" title="pagmo::nsga2"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::nsga2</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.nsga2.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.nsga2.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code> and printed to screen. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.nsga2" title="pygmo.nsga2"><code class="xref py py-class docutils literal"><span class="pre">nsga2</span></code></a>. A verbosity of <code class="docutils literal"><span class="pre">N</span></code> implies a log line each <code class="docutils literal"><span class="pre">N</span></code> generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">ideal_point</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">ideal_point</span></code> (1D numpy array), the ideal point of the current population (cropped to max 5 dimensions only in the screen output)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">nsga2</span><span class="p">(</span><span class="n">gen</span><span class="o">=</span><span class="mi">100</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">zdt</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="mi">40</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:        ideal1:        ideal2:</span>
<span class="go">   1              0      0.0033062        2.44966</span>
<span class="go">  21            800    0.000275601       0.893137</span>
<span class="go">  41           1600    3.15834e-05        0.44117</span>
<span class="go">  61           2400     2.3664e-05       0.206365</span>
<span class="go">  81           3200     2.3664e-05       0.133305</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">nsga2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 0, array([ 0.0033062 ,  2.44965599])), (21, 800, array([  2.75601086e-04 ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/nsga2.html#_CPPv2NK5pagmo5nsga27get_logEv" title="pagmo::nsga2::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::nsga2::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.nsga2.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.nsga2.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.mbh">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">mbh</code><span class="sig-paren">(</span><em>algo = compass_search()</em>, <em>stop = 5</em>, <em>perturb = 1e-2</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.mbh" title="Permalink to this definition">¶</a></dt>
<dd><p>Monotonic Basin Hopping (generalized).</p>
<p>Monotonic basin hopping, or simply, basin hopping, is an algorithm rooted in the idea of mapping
the objective function <span class="math">\(f(\mathbf x_0)\)</span> into the local minima found starting from <span class="math">\(\mathbf x_0\)</span>.
This simple idea allows a substantial increase of efficiency in solving problems, such as the Lennard-Jones
cluster or the MGA-1DSM interplanetary trajectory problem that are conjectured to have a so-called
funnel structure.</p>
<p>In pygmo we provide an original generalization of this concept resulting in a meta-algorithm that operates
on any <a class="reference internal" href="../py_population.html#pygmo.population" title="pygmo.population"><code class="xref py py-class docutils literal"><span class="pre">pygmo.population</span></code></a> using any suitable user-defined algorithm (UDA). When a population containing a single
individual is used and coupled with a local optimizer, the original method is recovered.
The pseudo code of our generalized version is:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>&gt; Select a pygmo population
&gt; Select a UDA
&gt; Store best individual
&gt; while i &lt; stop_criteria
&gt; &gt; Perturb the population in a selected neighbourhood
&gt; &gt; Evolve the population using the algorithm
&gt; &gt; if the best individual is improved
&gt; &gt; &gt; increment i
&gt; &gt; &gt; update best individual
&gt; &gt; else
&gt; &gt; &gt; i = 0
</pre></div>
</div>
<p><a class="reference internal" href="#pygmo.mbh" title="pygmo.mbh"><code class="xref py py-class docutils literal"><span class="pre">pygmo.mbh</span></code></a> is a user-defined algorithm (UDA) that can be used to construct <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">pygmo.algorithm</span></code></a> objects.</p>
<p>See: <a class="reference external" href="https://arxiv.org/pdf/cond-mat/9803344.pdf">https://arxiv.org/pdf/cond-mat/9803344.pdf</a> for the paper introducing the basin hopping idea for a Lennard-Jones
cluster optimization.</p>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/mbh.html#_CPPv2N5pagmo3mbhE" title="pagmo::mbh"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::mbh</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>algo</strong> – an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> or a user-defined algorithm (either C++ or Python - note that
<em>algo</em> will be deep-copied and stored inside the <a class="reference internal" href="#pygmo.mbh" title="pygmo.mbh"><code class="xref py py-class docutils literal"><span class="pre">mbh</span></code></a> instance)</li>
<li><strong>stop</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – consecutive runs of the inner algorithm that need to result in no improvement for
<a class="reference internal" href="#pygmo.mbh" title="pygmo.mbh"><code class="xref py py-class docutils literal"><span class="pre">mbh</span></code></a> to stop</li>
<li><strong>perturb</strong> (<code class="docutils literal"><span class="pre">float</span></code> or array-like object) – perturb the perturbation to be applied to each component</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>perturb</em> (or one of its components, if <em>perturb</em> is an array) is not in the
(0,1] range</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by the constructor of <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">pygmo.algorithm</span></code></a>, or by
failures at the intersection between C++ and Python (e.g., type conversion errors, mismatched function
signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pygmo.mbh.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.mbh.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code>. The log frequency depends on the verbosity parameter
(by default nothing is logged) which can be set calling <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on a <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> constructed
with an <a class="reference internal" href="#pygmo.mbh" title="pygmo.mbh"><code class="xref py py-class docutils literal"><span class="pre">mbh</span></code></a>. A verbosity level <code class="docutils literal"><span class="pre">N</span> <span class="pre">&gt;</span> <span class="pre">0</span></code> will log one line at the end of each call to the inner algorithm.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each call of the inner algorithm, the values <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">Violated</span></code>, <code class="docutils literal"><span class="pre">Viol.</span> <span class="pre">Norm</span></code> and <code class="docutils literal"><span class="pre">Trial</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), the number of fitness evaluations made</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the objective function of the best fitness currently in the population</li>
<li><code class="docutils literal"><span class="pre">Violated</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), the number of constraints currently violated by the best solution</li>
<li><code class="docutils literal"><span class="pre">Viol.</span> <span class="pre">Norm</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the norm of the violation (discounted already by the constraints tolerance)</li>
<li><code class="docutils literal"><span class="pre">Trial</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), the trial number (which will determine the algorithm stop)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">mbh</span><span class="p">(</span><span class="n">algorithm</span><span class="p">(</span><span class="n">de</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">10</span><span class="p">))))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">cec2013</span><span class="p">(</span><span class="n">prob_id</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">dim</span> <span class="o">=</span> <span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Fevals:          Best:      Violated:    Viol. Norm:         Trial:</span>
<span class="go">    440        25162.3              0              0              0</span>
<span class="go">    880          14318              0              0              0</span>
<span class="go">   1320        11178.2              0              0              0</span>
<span class="go">   1760        6613.71              0              0              0</span>
<span class="go">   2200        6613.71              0              0              1</span>
<span class="go">   2640        6124.62              0              0              0</span>
<span class="go">   3080        6124.62              0              0              1</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/mbh.html#_CPPv2NK5pagmo3mbh7get_logEv" title="pagmo::mbh::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::mbh::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.mbh.get_perturb">
<code class="descname">get_perturb</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.mbh.get_perturb" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the perturbation vector.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the perturbation vector</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">1D NumPy float array</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.mbh.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.mbh.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the seed value that was used for the construction of this <a class="reference internal" href="#pygmo.mbh" title="pygmo.mbh"><code class="xref py py-class docutils literal"><span class="pre">mbh</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the seed value</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.mbh.get_verbosity">
<code class="descname">get_verbosity</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.mbh.get_verbosity" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the verbosity level value that was used for the construction of this <a class="reference internal" href="#pygmo.mbh" title="pygmo.mbh"><code class="xref py py-class docutils literal"><span class="pre">mbh</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the verbosity level</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.mbh.inner_algorithm">
<code class="descname">inner_algorithm</code><a class="headerlink" href="#pygmo.mbh.inner_algorithm" title="Permalink to this definition">¶</a></dt>
<dd><p>Inner algorithm of the meta-algorithm</p>
<p>This read-only property gives direct access to the <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> stored within the meta-algorithm.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a reference to the inner algorithm</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.mbh.set_perturb">
<code class="descname">set_perturb</code><span class="sig-paren">(</span><em>perturb</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.mbh.set_perturb" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the perturbation vector.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>perturb</strong> (<em>array-like object</em>) – perturb the perturbation to be applied to each component</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g., type conversion errors,
mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.cstrs_self_adaptive">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">cstrs_self_adaptive</code><span class="sig-paren">(</span><em>iters = 1</em>, <em>algo = de()</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.cstrs_self_adaptive" title="Permalink to this definition">¶</a></dt>
<dd><p>This meta-algorithm implements a constraint handling technique that allows the use of any user-defined algorithm
(UDA) able to deal with single-objective unconstrained problems, on single-objective constrained problems. The
technique self-adapts its parameters during each successive call to the inner UDA basing its decisions on the entire
underlying population. The resulting approach is an alternative to using the meta-problem <a class="reference internal" href="../problems/py_problems.html#pygmo.unconstrain" title="pygmo.unconstrain"><code class="xref py py-class docutils literal"><span class="pre">unconstrain</span></code></a>
to transform the constrained fitness into an unconstrained fitness.</p>
<p>The self-adaptive constraints handling meta-algorithm is largely based on the ideas of Faramani and Wright but it
extends their use to any-algorithm, in particular to non generational, population based, evolutionary approaches where
a steady-state reinsertion is used (i.e. as soon as an individual is found fit it is immediately reinserted into the
population and will influence the next offspring genetic material).</p>
<p>Each decision vector is assigned an infeasibility measure <span class="math">\(\iota\)</span> which accounts for the normalized violation of
all the constraints (discounted by the constraints tolerance as returned by <a class="reference internal" href="../py_problem.html#pygmo.problem.c_tol" title="pygmo.problem.c_tol"><code class="xref py py-attr docutils literal"><span class="pre">pygmo.problem.c_tol</span></code></a>). The
normalization factor used <span class="math">\(c_{j_{max}}\)</span> is the maximum violation of the <span class="math">\(j\)</span> constraint.</p>
<p>As in the original paper, three individuals in the evolving population are then used to penalize the single
objective.</p>
<div class="math">
\[\begin{split}\begin{array}{rl}
\check X &amp; \mbox{: the best decision vector} \\
\hat X &amp; \mbox{: the worst decision vector} \\
\breve X &amp; \mbox{: the decision vector with the highest objective}
\end{array}\end{split}\]</div>
<p>The best and worst decision vectors are defined accounting for their infeasibilities and for the value of the
objective function. Using the above definitions the overall pseudo code can be summarized as follows:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>&gt; Select a pygmo.population (related to a single-objective constrained problem)
&gt; Select a UDA (able to solve single-objective unconstrained problems)
&gt; while i &lt; iter
&gt; &gt; Compute the normalization factors (will depend on the current population)
&gt; &gt; Compute the best, worst, highest (will depend on the current population)
&gt; &gt; Evolve the population using the UDA and a penalized objective
&gt; &gt; Reinsert the best decision vector from the previous evolution
</pre></div>
</div>
<p><a class="reference internal" href="#pygmo.cstrs_self_adaptive" title="pygmo.cstrs_self_adaptive"><code class="xref py py-class docutils literal"><span class="pre">pygmo.cstrs_self_adaptive</span></code></a> is a user-defined algorithm (UDA) that can be used to construct <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">pygmo.algorithm</span></code></a> objects.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Self-adaptive constraints handling implements an internal cache to avoid the re-evaluation of the fitness
for decision vectors already evaluated. This makes the final counter of fitness evaluations somewhat unpredictable.
The number of function evaluation will be bounded to <em>iters</em> times the fevals made by one call to the inner UDA. The
internal cache is reset at each iteration, but its size will grow unlimited during each call to
the inner UDA evolve method.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Several modification were made to the original Faramani and Wright ideas to allow their approach to work on
corner cases and with any UDAs. Most notably, a violation to the <span class="math">\(j\)</span>-th  constraint is ignored if all
the decision vectors in the population satisfy that particular constraint (i.e. if <span class="math">\(c_{j_{max}} = 0\)</span>).</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The performances of <a class="reference internal" href="#pygmo.cstrs_self_adaptive" title="pygmo.cstrs_self_adaptive"><code class="xref py py-class docutils literal"><span class="pre">cstrs_self_adaptive</span></code></a> are highly dependent on the particular inner
algorithm employed and in particular to its parameters (generations / iterations).</p>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">Farmani, Raziyeh, and Jonathan A. Wright. “Self-adaptive fitness formulation for constrained optimization.” IEEE
Transactions on Evolutionary Computation 7.5 (2003): 445-455.</p>
</div>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/cstrs_self_adaptive.html#_CPPv2N5pagmo19cstrs_self_adaptiveE" title="pagmo::cstrs_self_adaptive"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::cstrs_self_adaptive</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>iter</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of iterations (i.e. calls to the innel algorithm evolve)</li>
<li><strong>algo</strong> – an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> or a user-defined algorithm (either C++ or Python - note that <em>algo</em>
will be deep-copied and stored inside the <a class="reference internal" href="#pygmo.cstrs_self_adaptive" title="pygmo.cstrs_self_adaptive"><code class="xref py py-class docutils literal"><span class="pre">cstrs_self_adaptive</span></code></a> instance)</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>iters</em> is negative or greater than an implementation-defined value</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by the constructor of <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">pygmo.algorithm</span></code></a>, or by
failures at the intersection between C++ and Python (e.g., type conversion errors, mismatched function
signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pygmo.cstrs_self_adaptive.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.cstrs_self_adaptive.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code>. The log frequency depends on the verbosity parameter
(by default nothing is logged) which can be set calling <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on a <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> constructed
with an <a class="reference internal" href="#pygmo.cstrs_self_adaptive" title="pygmo.cstrs_self_adaptive"><code class="xref py py-class docutils literal"><span class="pre">cstrs_self_adaptive</span></code></a>. A verbosity level of <code class="docutils literal"><span class="pre">N</span> <span class="pre">&gt;</span> <span class="pre">0</span></code> will log one line each <code class="docutils literal"><span class="pre">N</span></code> <code class="docutils literal"><span class="pre">iters</span></code>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each call of the inner algorithm, the values <code class="docutils literal"><span class="pre">Iters</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">Infeasibility</span></code>,
<code class="docutils literal"><span class="pre">Violated</span></code>, <code class="docutils literal"><span class="pre">Viol.</span> <span class="pre">Norm</span></code> and <code class="docutils literal"><span class="pre">N.</span> <span class="pre">Feasible</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Iters</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), the number of iterations made (i.e. calls to the evolve method of the inner algorithm)</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), the number of fitness evaluations made</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the objective function of the best fitness currently in the population</li>
<li><code class="docutils literal"><span class="pre">Infeasibility</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the aggregated (and normalized) infeasibility value of <code class="docutils literal"><span class="pre">Best</span></code></li>
<li><code class="docutils literal"><span class="pre">Violated</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), the number of constraints currently violated by the best solution</li>
<li><code class="docutils literal"><span class="pre">Viol.</span> <span class="pre">Norm</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the norm of the violation (discounted already by the constraints tolerance)</li>
<li><code class="docutils literal"><span class="pre">N.</span> <span class="pre">Feasible</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), the number of feasible individuals currently in the population.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">cstrs_self_adaptive</span><span class="p">(</span><span class="n">iters</span> <span class="o">=</span> <span class="mi">20</span><span class="p">,</span> <span class="n">algo</span> <span class="o">=</span> <span class="n">de</span><span class="p">(</span><span class="mi">10</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">cec2006</span><span class="p">(</span><span class="n">prob_id</span> <span class="o">=</span> <span class="mi">1</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Iter:        Fevals:          Best: Infeasibility:      Violated:    Viol. Norm:   N. Feasible:</span>
<span class="go">    1              0       -96.5435        0.34607              4        177.705              0 i</span>
<span class="go">    4            600       -96.5435       0.360913              4        177.705              0 i</span>
<span class="go">    7           1200       -96.5435        0.36434              4        177.705              0 i</span>
<span class="go">   10           1800       -96.5435       0.362307              4        177.705              0 i</span>
<span class="go">   13           2400       -23.2502       0.098049              4        37.1092              0 i</span>
<span class="go">   16           3000       -23.2502       0.071571              4        37.1092              0 i</span>
<span class="go">   19           3600       -23.2502       0.257604              4        37.1092              0 i</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">moead</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 0, -96.54346700540063, 0.34606950943401493, 4, 177.70482046341274, 0), (4, 600, ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/cstrs_self_adaptive.html#_CPPv2NK5pagmo19cstrs_self_adaptive7get_logEv" title="pagmo::cstrs_self_adaptive::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::cstrs_self_adaptive::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.cstrs_self_adaptive.inner_algorithm">
<code class="descname">inner_algorithm</code><a class="headerlink" href="#pygmo.cstrs_self_adaptive.inner_algorithm" title="Permalink to this definition">¶</a></dt>
<dd><p>Inner algorithm of the meta-algorithm</p>
<p>This read-only property gives direct access to the <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> stored within the meta-algorithm.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a reference to the inner algorithm</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.nlopt">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">nlopt</code><span class="sig-paren">(</span><em>solver = &quot;cobyla&quot;</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.nlopt" title="Permalink to this definition">¶</a></dt>
<dd><p>NLopt algorithms.</p>
<p>This user-defined algorithm wraps a selection of solvers from the
<a class="reference external" href="https://nlopt.readthedocs.io/en/latest/">NLopt</a> library, focusing on
local optimisation (both gradient-based and derivative-free). The complete list of supported
NLopt algorithms is:</p>
<ul class="simple">
<li>COBYLA,</li>
<li>BOBYQA,</li>
<li>NEWUOA + bound constraints,</li>
<li>PRAXIS,</li>
<li>Nelder-Mead simplex,</li>
<li>sbplx,</li>
<li>MMA (Method of Moving Asymptotes),</li>
<li>CCSA,</li>
<li>SLSQP,</li>
<li>low-storage BFGS,</li>
<li>preconditioned truncated Newton,</li>
<li>shifted limited-memory variable-metric,</li>
<li>augmented Lagrangian algorithm.</li>
</ul>
<p>The desired NLopt solver is selected upon construction of an <a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">nlopt</span></code></a> algorithm. Various properties
of the solver (e.g., the stopping criteria) can be configured via class attributes. Multiple
stopping criteria can be active at the same time: the optimisation will stop as soon as at least one stopping criterion
is satisfied. By default, only the <code class="docutils literal"><span class="pre">xtol_rel</span></code> stopping criterion is active (see <a class="reference internal" href="#pygmo.nlopt.xtol_rel" title="pygmo.nlopt.xtol_rel"><code class="xref py py-attr docutils literal"><span class="pre">xtol_rel</span></code></a>).</p>
<p>All NLopt solvers support only single-objective optimisation, and, as usual in pygmo, minimisation
is always assumed. The gradient-based algorithms require the optimisation problem to provide a gradient.
Some solvers support equality and/or inequality constraints. The constraints’ tolerances will
be set to those specified in the <a class="reference internal" href="../py_problem.html#pygmo.problem" title="pygmo.problem"><code class="xref py py-class docutils literal"><span class="pre">problem</span></code></a> being optimised (see <a class="reference internal" href="../py_problem.html#pygmo.problem.c_tol" title="pygmo.problem.c_tol"><code class="xref py py-attr docutils literal"><span class="pre">pygmo.problem.c_tol</span></code></a>).</p>
<p>In order to support pygmo’s population-based optimisation model, the <code class="docutils literal"><span class="pre">evolve()</span></code> method will select
a single individual from the input <a class="reference internal" href="../py_population.html#pygmo.population" title="pygmo.population"><code class="xref py py-class docutils literal"><span class="pre">population</span></code></a> to be optimised by the NLopt solver.
If the optimisation produces a better individual (as established by <a class="reference internal" href="../utils/py_con_utils.html#pygmo.compare_fc" title="pygmo.compare_fc"><code class="xref py py-func docutils literal"><span class="pre">compare_fc()</span></code></a>),
the optimised individual will be inserted back into the population.
The selection and replacement strategies can be configured via the <a class="reference internal" href="#pygmo.nlopt.selection" title="pygmo.nlopt.selection"><code class="xref py py-attr docutils literal"><span class="pre">selection</span></code></a>
and <a class="reference internal" href="#pygmo.nlopt.replacement" title="pygmo.nlopt.replacement"><code class="xref py py-attr docutils literal"><span class="pre">replacement</span></code></a> attributes.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This user-defined algorithm is available only if pygmo was compiled with the <code class="docutils literal"><span class="pre">PAGMO_WITH_NLOPT</span></code> option
enabled (see the <a class="reference internal" href="../../../install.html#install"><span class="std std-ref">installation instructions</span></a>).</p>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">The <a class="reference external" href="http://nlopt.readthedocs.io/en/latest/NLopt_Algorithms/">NLopt website</a> contains a detailed description
of each supported solver.</p>
</div>
<p>This constructor will initialise an <a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">nlopt</span></code></a> object which will use the NLopt algorithm specified by
the input string <em>solver</em>, the <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> individual selection strategy and the <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> individual
replacement strategy. <em>solver</em> is translated to an NLopt algorithm type according to the following
translation table:</p>
<table border="1" class="docutils">
<colgroup>
<col width="47%" />
<col width="53%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head"><em>solver</em> string</th>
<th class="head">NLopt algorithm</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;cobyla&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LN_COBYLA</span></code></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">&quot;bobyqa&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LN_BOBYQA</span></code></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;newuoa&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LN_NEWUOA</span></code></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">&quot;newuoa_bound&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LN_NEWUOA_BOUND</span></code></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;praxis&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LN_PRAXIS</span></code></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">&quot;neldermead&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LN_NELDERMEAD</span></code></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;sbplx&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LN_SBPLX</span></code></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">&quot;mma&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_MMA</span></code></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;ccsaq&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_CCSAQ</span></code></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">&quot;slsqp&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_SLSQP</span></code></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;lbfgs&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_LBFGS</span></code></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">&quot;tnewton_precond_restart&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_TNEWTON_PRECOND_RESTART</span></code></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;tnewton_precond&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_TNEWTON_PRECOND</span></code></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">&quot;tnewton_restart&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_TNEWTON_RESTART</span></code></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;tnewton&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_TNEWTON</span></code></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">&quot;var2&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_VAR2</span></code></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;var1&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_LD_VAR1</span></code></td>
</tr>
<tr class="row-odd"><td><code class="docutils literal"><span class="pre">&quot;auglag&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_AUGLAG</span></code></td>
</tr>
<tr class="row-even"><td><code class="docutils literal"><span class="pre">&quot;auglag_eq&quot;</span></code></td>
<td><code class="docutils literal"><span class="pre">NLOPT_AUGLAG_EQ</span></code></td>
</tr>
</tbody>
</table>
<p>The parameters of the selected solver can be configured via the attributes of this class.</p>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/nlopt.html#_CPPv2N5pagmo5nloptE" title="pagmo::nlopt"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::nlopt</span></code></a>.</p>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last">The <a class="reference external" href="http://nlopt.readthedocs.io/en/latest/NLopt_Algorithms/">NLopt website</a> contains a detailed
description of each supported solver.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>solver</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – the name of the NLopt algorithm that will be used by this <a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">nlopt</span></code></a> object</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#RuntimeError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">RuntimeError</span></code></a> – if the NLopt version is not at least 2</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>solver</em> is not one of the allowed algorithm names</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nl</span> <span class="o">=</span> <span class="n">nlopt</span><span class="p">(</span><span class="s1">&#39;slsqp&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nl</span><span class="o">.</span><span class="n">xtol_rel</span> <span class="o">=</span> <span class="mf">1E-6</span> <span class="c1"># Change the default value of the xtol_rel stopping criterion</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">nl</span><span class="o">.</span><span class="n">xtol_rel</span> 
<span class="go">1E-6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">nl</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">luksan_vlcek1</span><span class="p">(</span><span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span><span class="o">.</span><span class="n">c_tol</span> <span class="o">=</span> <span class="p">[</span><span class="mf">1E-6</span><span class="p">]</span> <span class="o">*</span> <span class="mi">18</span> <span class="c1"># Set constraints tolerance to 1E-6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">   objevals:       objval:      violated:    viol. norm:</span>
<span class="go">           1       95959.4             18        538.227 i</span>
<span class="go">           2       89282.7             18        5177.42 i</span>
<span class="go">           3         75580             18        464.206 i</span>
<span class="go">           4         75580             18        464.206 i</span>
<span class="go">           5       77737.6             18        1095.94 i</span>
<span class="go">           6         41162             18        350.446 i</span>
<span class="go">           7         41162             18        350.446 i</span>
<span class="go">           8         67881             18        362.454 i</span>
<span class="go">           9       30502.2             18        249.762 i</span>
<span class="go">          10       30502.2             18        249.762 i</span>
<span class="go">          11       7266.73             18        95.5946 i</span>
<span class="go">          12        4510.3             18        42.2385 i</span>
<span class="go">          13       2400.66             18        35.2507 i</span>
<span class="go">          14       34051.9             18        749.355 i</span>
<span class="go">          15       1657.41             18        32.1575 i</span>
<span class="go">          16       1657.41             18        32.1575 i</span>
<span class="go">          17       1564.44             18        12.5042 i</span>
<span class="go">          18       275.987             14        6.22676 i</span>
<span class="go">          19       232.765             12         12.442 i</span>
<span class="go">          20       161.892             15        4.00744 i</span>
<span class="go">          21       161.892             15        4.00744 i</span>
<span class="go">          22       17.6821             11        1.78909 i</span>
<span class="go">          23       7.71103              5       0.130386 i</span>
<span class="go">          24       6.24758              4     0.00736759 i</span>
<span class="go">          25       6.23325              1    5.12547e-05 i</span>
<span class="go">          26        6.2325              0              0</span>
<span class="go">          27       6.23246              0              0</span>
<span class="go">          28       6.23246              0              0</span>
<span class="go">          29       6.23246              0              0</span>
<span class="go">          30       6.23246              0              0</span>

<span class="go">Optimisation return status: NLOPT_XTOL_REACHED (value = 4, Optimization stopped because xtol_rel or xtol_abs was reached)</span>
</pre></div>
</div>
<dl class="attribute">
<dt id="pygmo.nlopt.ftol_abs">
<code class="descname">ftol_abs</code><a class="headerlink" href="#pygmo.nlopt.ftol_abs" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">ftol_abs</span></code> stopping criterion.</p>
<p>The <code class="docutils literal"><span class="pre">ftol_abs</span></code> stopping criterion instructs the solver to stop when an optimization step
(or an estimate of the optimum) changes the function value by less than <code class="docutils literal"><span class="pre">ftol_abs</span></code>.
Defaults to 0 (that is, this stopping criterion is disabled by default).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the value of the <code class="docutils literal"><span class="pre">ftol_abs</span></code> stopping criterion</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">float</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if, when setting this property, a <code class="docutils literal"><span class="pre">NaN</span></code> is passed</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.nlopt.ftol_rel">
<code class="descname">ftol_rel</code><a class="headerlink" href="#pygmo.nlopt.ftol_rel" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">ftol_rel</span></code> stopping criterion.</p>
<p>The <code class="docutils literal"><span class="pre">ftol_rel</span></code> stopping criterion instructs the solver to stop when an optimization step (or an estimate of the
optimum) changes the objective function value by less than <code class="docutils literal"><span class="pre">ftol_rel</span></code> multiplied by the absolute value of the
function value. Defaults to 0 (that is, this stopping criterion is disabled by default).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the value of the <code class="docutils literal"><span class="pre">ftol_rel</span></code> stopping criterion</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">float</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if, when setting this property, a <code class="docutils literal"><span class="pre">NaN</span></code> is passed</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.nlopt.get_last_opt_result">
<code class="descname">get_last_opt_result</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.nlopt.get_last_opt_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the result of the last optimisation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the NLopt return code for the last optimisation run, or <code class="docutils literal"><span class="pre">NLOPT_SUCCESS</span></code> if no optimisations have been run yet</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.nlopt.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.nlopt.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Optimisation log.</p>
<p>The optimisation log is a collection of log data lines. A log data line is a tuple consisting of:</p>
<ul class="simple">
<li>the number of objective function evaluations made so far,</li>
<li>the objective function value for the current decision vector,</li>
<li>the number of constraints violated by the current decision vector,</li>
<li>the constraints violation norm for the current decision vector,</li>
<li>a boolean flag signalling the feasibility of the current decision vector.</li>
</ul>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the optimisation log</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.nlopt.get_solver_name">
<code class="descname">get_solver_name</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.nlopt.get_solver_name" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the name of the NLopt solver used during construction.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the name of the NLopt solver used during construction</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">str</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.nlopt.local_optimizer">
<code class="descname">local_optimizer</code><a class="headerlink" href="#pygmo.nlopt.local_optimizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Local optimizer.</p>
<p>Some NLopt algorithms rely on other NLopt algorithms as local/subsidiary optimizers.
This property, of type <a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">nlopt</span></code></a>, allows to set such local optimizer.
By default, no local optimizer is specified, and the property is set to <code class="docutils literal"><span class="pre">None</span></code>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">At the present time, only the <code class="docutils literal"><span class="pre">&quot;auglag&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;auglag_eq&quot;</span></code> solvers make use
of a local optimizer. Setting a local optimizer on any other solver will have no effect.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The objective function, bounds, and nonlinear-constraint parameters of the local
optimizer are ignored (as they are provided by the parent optimizer). Conversely, the stopping
criteria should be specified in the local optimizer.The verbosity of
the local optimizer is also forcibly set to zero during the optimisation.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the local optimizer, or <code class="docutils literal"><span class="pre">None</span></code> if not set</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#pygmo.nlopt" title="pygmo.nlopt"><code class="xref py py-class docutils literal"><span class="pre">nlopt</span></code></a></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python
(e.g., type conversion errors, mismatched function signatures, etc.), when setting the property</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.nlopt.maxeval">
<code class="descname">maxeval</code><a class="headerlink" href="#pygmo.nlopt.maxeval" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">maxeval</span></code> stopping criterion.</p>
<p>The <code class="docutils literal"><span class="pre">maxeval</span></code> stopping criterion instructs the solver to stop when the number of function evaluations exceeds
<code class="docutils literal"><span class="pre">maxeval</span></code>. Defaults to 0 (that is, this stopping criterion is disabled by default).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the value of the <code class="docutils literal"><span class="pre">maxeval</span></code> stopping criterion</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.nlopt.maxtime">
<code class="descname">maxtime</code><a class="headerlink" href="#pygmo.nlopt.maxtime" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">maxtime</span></code> stopping criterion.</p>
<p>The <code class="docutils literal"><span class="pre">maxtime</span></code> stopping criterion instructs the solver to stop when the optimization time (in seconds) exceeds
<code class="docutils literal"><span class="pre">maxtime</span></code>. Defaults to 0 (that is, this stopping criterion is disabled by default).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the value of the <code class="docutils literal"><span class="pre">maxtime</span></code> stopping criterion</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.nlopt.replacement">
<code class="descname">replacement</code><a class="headerlink" href="#pygmo.nlopt.replacement" title="Permalink to this definition">¶</a></dt>
<dd><p>Individual replacement policy.</p>
<p>This attribute represents the policy that is used in the <code class="docutils literal"><span class="pre">evolve()</span></code> method to select the individual
that will be replaced by the optimised individual. The attribute can be either a string or an integral.</p>
<p>If the attribute is a string, it must be one of <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> will select the best individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> will select the worst individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> will randomly choose one individual in the population.</li>
</ul>
<p><a class="reference internal" href="#pygmo.nlopt.set_random_sr_seed" title="pygmo.nlopt.set_random_sr_seed"><code class="xref py py-func docutils literal"><span class="pre">set_random_sr_seed()</span></code></a> can be used to seed the random number generator
used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> policy.</p>
<p>If the attribute is an integer, it represents the index (in the population) of the individual that will be
replaced by the optimised individual.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the individual replacement policy or index</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">str</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if the attribute is set to an invalid string</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code></a> – if the attribute is set to a value of an invalid type</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.nlopt.selection">
<code class="descname">selection</code><a class="headerlink" href="#pygmo.nlopt.selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Individual selection policy.</p>
<p>This attribute represents the policy that is used in the <code class="docutils literal"><span class="pre">evolve()</span></code> method to select the individual
that will be optimised. The attribute can be either a string or an integral.</p>
<p>If the attribute is a string, it must be one of <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> will select the best individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> will select the worst individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> will randomly choose one individual in the population.</li>
</ul>
<p><a class="reference internal" href="#pygmo.nlopt.set_random_sr_seed" title="pygmo.nlopt.set_random_sr_seed"><code class="xref py py-func docutils literal"><span class="pre">set_random_sr_seed()</span></code></a> can be used to seed the random number generator
used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> policy.</p>
<p>If the attribute is an integer, it represents the index (in the population) of the individual that is selected
for optimisation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the individual selection policy or index</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">str</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if the attribute is set to an invalid string</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code></a> – if the attribute is set to a value of an invalid type</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.nlopt.set_random_sr_seed">
<code class="descname">set_random_sr_seed</code><span class="sig-paren">(</span><em>seed</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.nlopt.set_random_sr_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the seed for the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> selection/replacement policies.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – the value that will be used to seed the random number generator used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>
election/replacement policies (see <a class="reference internal" href="#pygmo.nlopt.selection" title="pygmo.nlopt.selection"><code class="xref py py-attr docutils literal"><span class="pre">selection</span></code></a> and
<a class="reference internal" href="#pygmo.nlopt.replacement" title="pygmo.nlopt.replacement"><code class="xref py py-attr docutils literal"><span class="pre">replacement</span></code></a>)</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.nlopt.stopval">
<code class="descname">stopval</code><a class="headerlink" href="#pygmo.nlopt.stopval" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">stopval</span></code> stopping criterion.</p>
<p>The <code class="docutils literal"><span class="pre">stopval</span></code> stopping criterion instructs the solver to stop when an objective value less than
or equal to <code class="docutils literal"><span class="pre">stopval</span></code> is found. Defaults to the C constant <code class="docutils literal"><span class="pre">-HUGE_VAL</span></code> (that is, this stopping criterion
is disabled by default).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the value of the <code class="docutils literal"><span class="pre">stopval</span></code> stopping criterion</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">float</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if, when setting this property, a <code class="docutils literal"><span class="pre">NaN</span></code> is passed</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.nlopt.xtol_abs">
<code class="descname">xtol_abs</code><a class="headerlink" href="#pygmo.nlopt.xtol_abs" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">xtol_abs</span></code> stopping criterion.</p>
<p>The <code class="docutils literal"><span class="pre">xtol_abs</span></code> stopping criterion instructs the solver to stop when an optimization step (or an estimate of the
optimum) changes every parameter by less than <code class="docutils literal"><span class="pre">xtol_abs</span></code>. Defaults to 0 (that is, this stopping criterion is disabled
by default).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the value of the <code class="docutils literal"><span class="pre">xtol_abs</span></code> stopping criterion</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">float</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if, when setting this property, a <code class="docutils literal"><span class="pre">NaN</span></code> is passed</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.nlopt.xtol_rel">
<code class="descname">xtol_rel</code><a class="headerlink" href="#pygmo.nlopt.xtol_rel" title="Permalink to this definition">¶</a></dt>
<dd><p><code class="docutils literal"><span class="pre">xtol_rel</span></code> stopping criterion.</p>
<p>The <code class="docutils literal"><span class="pre">xtol_rel</span></code> stopping criterion instructs the solver to stop when an optimization step (or an estimate of the
optimum) changes every parameter by less than <code class="docutils literal"><span class="pre">xtol_rel</span></code> multiplied by the absolute value of the parameter.
Defaults to 1E-8.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the value of the <code class="docutils literal"><span class="pre">xtol_rel</span></code> stopping criterion</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">float</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if, when setting this property, a <code class="docutils literal"><span class="pre">NaN</span></code> is passed</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.ipopt">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">ipopt</code><a class="headerlink" href="#pygmo.ipopt" title="Permalink to this definition">¶</a></dt>
<dd><p>Ipopt.</p>
<div class="versionadded">
<p><span class="versionmodified">New in version 2.2.</span></p>
</div>
<p>This class is a user-defined algorithm (UDA) that wraps the Ipopt (Interior Point OPTimizer) solver,
a software package for large-scale nonlinear optimization. Ipopt is a powerful solver that
is able to handle robustly and efficiently constrained nonlinear opimization problems at high dimensionalities.</p>
<p>Ipopt supports only single-objective minimisation, and it requires the availability of the gradient in the
optimisation problem. If possible, for best results the Hessians should be provided as well (but Ipopt
can estimate numerically the Hessians if needed).</p>
<p>In order to support pygmo’s population-based optimisation model, the <code class="docutils literal"><span class="pre">evolve()</span></code> method will select
a single individual from the input <a class="reference internal" href="../py_population.html#pygmo.population" title="pygmo.population"><code class="xref py py-class docutils literal"><span class="pre">population</span></code></a> to be optimised.
If the optimisation produces a better individual (as established by <a class="reference internal" href="../utils/py_con_utils.html#pygmo.compare_fc" title="pygmo.compare_fc"><code class="xref py py-func docutils literal"><span class="pre">compare_fc()</span></code></a>),
the optimised individual will be inserted back into the population. The selection and replacement strategies
can be configured via the <a class="reference internal" href="#pygmo.ipopt.selection" title="pygmo.ipopt.selection"><code class="xref py py-attr docutils literal"><span class="pre">selection</span></code></a> and <a class="reference internal" href="#pygmo.ipopt.replacement" title="pygmo.ipopt.replacement"><code class="xref py py-attr docutils literal"><span class="pre">replacement</span></code></a> attributes.</p>
<p>Ipopt supports a large amount of options for the configuration of the optimisation run. The options
are divided into three categories:</p>
<ul class="simple">
<li><em>string</em> options (i.e., the type of the option is <code class="docutils literal"><span class="pre">str</span></code>),</li>
<li><em>integer</em> options (i.e., the type of the option is <code class="docutils literal"><span class="pre">int</span></code>),</li>
<li><em>numeric</em> options (i.e., the type of the option is <code class="docutils literal"><span class="pre">float</span></code>).</li>
</ul>
<p>The full list of options is available on the <a class="reference external" href="https://www.coin-or.org/Ipopt/documentation/node40.html">Ipopt website</a>.
<a class="reference internal" href="#pygmo.ipopt" title="pygmo.ipopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.ipopt</span></code></a> allows to configure any Ipopt option via methods such as <a class="reference internal" href="#pygmo.ipopt.set_string_options" title="pygmo.ipopt.set_string_options"><code class="xref py py-func docutils literal"><span class="pre">set_string_options()</span></code></a>,
<a class="reference internal" href="#pygmo.ipopt.set_string_option" title="pygmo.ipopt.set_string_option"><code class="xref py py-func docutils literal"><span class="pre">set_string_option()</span></code></a>, <a class="reference internal" href="#pygmo.ipopt.set_integer_options" title="pygmo.ipopt.set_integer_options"><code class="xref py py-func docutils literal"><span class="pre">set_integer_options()</span></code></a>, etc., which need to be used before
invoking the <code class="docutils literal"><span class="pre">evolve()</span></code> method.</p>
<p>If the user does not set any option, <a class="reference internal" href="#pygmo.ipopt" title="pygmo.ipopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.ipopt</span></code></a> use Ipopt’s default values for the options (see the
<a class="reference external" href="https://www.coin-or.org/Ipopt/documentation/node40.html">documentation</a>), with the following
modifications:</p>
<ul class="simple">
<li>if the <code class="docutils literal"><span class="pre">&quot;print_level&quot;</span></code> integer option is <strong>not</strong> set by the user, it will be set to 0 by <a class="reference internal" href="#pygmo.ipopt" title="pygmo.ipopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.ipopt</span></code></a> (this will
suppress most screen output produced by the solver - note that we support an alternative form of logging via
the <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">pygmo.algorithm.set_verbosity()</span></code></a> machinery);</li>
<li>if the <code class="docutils literal"><span class="pre">&quot;hessian_approximation&quot;</span></code> string option is <strong>not</strong> set by the user and the optimisation problem does
<strong>not</strong> provide the Hessians, then the option will be set to <code class="docutils literal"><span class="pre">&quot;limited-memory&quot;</span></code> by <a class="reference internal" href="#pygmo.ipopt" title="pygmo.ipopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.ipopt</span></code></a>. This makes it
possible to optimise problems without Hessians out-of-the-box (i.e., Ipopt will approximate numerically the
Hessians for you);</li>
<li>if the <code class="docutils literal"><span class="pre">&quot;constr_viol_tol&quot;</span></code> numeric option is <strong>not</strong> set by the user and the optimisation problem is constrained,
then <a class="reference internal" href="#pygmo.ipopt" title="pygmo.ipopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.ipopt</span></code></a> will compute the minimum value <code class="docutils literal"><span class="pre">min_tol</span></code> in the vector returned by <a class="reference internal" href="../py_problem.html#pygmo.problem.c_tol" title="pygmo.problem.c_tol"><code class="xref py py-attr docutils literal"><span class="pre">pygmo.problem.c_tol</span></code></a>
for the optimisation problem at hand. If <code class="docutils literal"><span class="pre">min_tol</span></code> is nonzero, then the <code class="docutils literal"><span class="pre">&quot;constr_viol_tol&quot;</span></code> Ipopt option will
be set to <code class="docutils literal"><span class="pre">min_tol</span></code>, otherwise the default Ipopt value (1E-4) will be used for the option. This ensures that,
if the constraint tolerance is not explicitly set by the user, a solution deemed feasible by Ipopt is also
deemed feasible by pygmo (but the opposite is not necessarily true).</li>
</ul>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">This user-defined algorithm is available only if pygmo was compiled with the <code class="docutils literal"><span class="pre">PAGMO_WITH_IPOPT</span></code> option
enabled (see the <a class="reference internal" href="../../../install.html#install"><span class="std std-ref">installation instructions</span></a>).</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Ipopt is not thread-safe, and thus it cannot be used in a <a class="reference internal" href="../islands/py_islands.html#pygmo.thread_island" title="pygmo.thread_island"><code class="xref py py-class docutils literal"><span class="pre">pygmo.thread_island</span></code></a>.</p>
</div>
<div class="admonition seealso">
<p class="first admonition-title">See also</p>
<p class="last"><a class="reference external" href="https://projects.coin-or.org/Ipopt">https://projects.coin-or.org/Ipopt</a>.</p>
</div>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/ipopt.html#_CPPv2N5pagmo5ipoptE" title="pagmo::ipopt"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::ipopt</span></code></a>.</p>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_numeric_option</span><span class="p">(</span><span class="s2">&quot;tol&quot;</span><span class="p">,</span><span class="mf">1E-9</span><span class="p">)</span> <span class="c1"># Change the relative convergence tolerance</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_numeric_options</span><span class="p">()</span> 
<span class="go">{&#39;tol&#39;: 1e-09}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">ip</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">luksan_vlcek1</span><span class="p">(</span><span class="mi">20</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span><span class="o">.</span><span class="n">c_tol</span> <span class="o">=</span> <span class="p">[</span><span class="mf">1E-6</span><span class="p">]</span> <span class="o">*</span> <span class="mi">18</span> <span class="c1"># Set constraints tolerance to 1E-6</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 

<span class="go">******************************************************************************</span>
<span class="go">This program contains Ipopt, a library for large-scale nonlinear optimization.</span>
<span class="go">Ipopt is released as open source code under the Eclipse Public License (EPL).</span>
<span class="go">        For more information visit http://projects.coin-or.org/Ipopt</span>
<span class="go">******************************************************************************</span>


<span class="go">objevals:        objval:      violated:    viol. norm:</span>
<span class="go">        1         201174             18         1075.3 i</span>
<span class="go">        2         209320             18        691.814 i</span>
<span class="go">        3        36222.3             18        341.639 i</span>
<span class="go">        4        11158.1             18        121.097 i</span>
<span class="go">        5        4270.38             18        46.4742 i</span>
<span class="go">        6        2054.03             18        20.7306 i</span>
<span class="go">        7        705.959             18        5.43118 i</span>
<span class="go">        8        37.8304             18        1.52099 i</span>
<span class="go">        9        2.89066             12       0.128862 i</span>
<span class="go">       10       0.300807              3      0.0165902 i</span>
<span class="go">       11     0.00430279              3    0.000496496 i</span>
<span class="go">       12    7.54121e-06              2    9.70735e-06 i</span>
<span class="go">       13    4.34249e-08              0              0</span>
<span class="go">       14    3.71925e-10              0              0</span>
<span class="go">       15    3.54406e-13              0              0</span>
<span class="go">       16    2.37071e-18              0              0</span>

<span class="go">Optimisation return status: Solve_Succeeded (value = 0)</span>
</pre></div>
</div>
<dl class="method">
<dt id="pygmo.ipopt.get_integer_options">
<code class="descname">get_integer_options</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.get_integer_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Get integer options.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a name-value dictionary of optimisation integer options</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">dict</span></code> of <code class="docutils literal"><span class="pre">str</span></code>-<code class="docutils literal"><span class="pre">int</span></code> pairs</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_integer_option</span><span class="p">(</span><span class="s2">&quot;print_level&quot;</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_integer_options</span><span class="p">()</span>
<span class="go">{&#39;print_level&#39;: 3}</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.get_last_opt_result">
<code class="descname">get_last_opt_result</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.get_last_opt_result" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the result of the last optimisation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the Ipopt return code for the last optimisation run, or <code class="docutils literal"><span class="pre">Ipopt::Solve_Succeeded</span></code> if no optimisations have been run yet</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_last_opt_result</span><span class="p">()</span>
<span class="go">0</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Optimisation log.</p>
<p>The optimisation log is a collection of log data lines. A log data line is a tuple consisting of:</p>
<ul class="simple">
<li>the number of objective function evaluations made so far,</li>
<li>the objective function value for the current decision vector,</li>
<li>the number of constraints violated by the current decision vector,</li>
<li>the constraints violation norm for the current decision vector,</li>
<li>a boolean flag signalling the feasibility of the current decision vector.</li>
</ul>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the optimisation log</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code></td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The number of constraints violated, the constraints violation norm and the feasibility flag stored in the log
are all determined via the facilities and the tolerances specified within <a class="reference internal" href="../py_problem.html#pygmo.problem" title="pygmo.problem"><code class="xref py py-class docutils literal"><span class="pre">pygmo.problem</span></code></a>. That
is, they might not necessarily be consistent with Ipopt’s notion of feasibility. See the explanation
of how the <code class="docutils literal"><span class="pre">&quot;constr_viol_tol&quot;</span></code> numeric option is handled in <a class="reference internal" href="#pygmo.ipopt" title="pygmo.ipopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.ipopt</span></code></a>.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Ipopt supports its own logging format and protocol, including the ability to print to screen and write to file.
Ipopt’s screen logging is disabled by default (i.e., the Ipopt verbosity setting is set to 0 - see
<a class="reference internal" href="#pygmo.ipopt" title="pygmo.ipopt"><code class="xref py py-class docutils literal"><span class="pre">pygmo.ipopt</span></code></a>). On-screen logging can be enabled via the <code class="docutils literal"><span class="pre">&quot;print_level&quot;</span></code> string option.</p>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.get_numeric_options">
<code class="descname">get_numeric_options</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.get_numeric_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Get numeric options.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a name-value dictionary of optimisation numeric options</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">dict</span></code> of <code class="docutils literal"><span class="pre">str</span></code>-<code class="docutils literal"><span class="pre">float</span></code> pairs</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_numeric_option</span><span class="p">(</span><span class="s2">&quot;tol&quot;</span><span class="p">,</span><span class="mf">1E-4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_numeric_options</span><span class="p">()</span> 
<span class="go">{&#39;tol&#39;: 1E-4}</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.get_string_options">
<code class="descname">get_string_options</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.get_string_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Get string options.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">a name-value dictionary of optimisation string options</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">dict</span></code> of <code class="docutils literal"><span class="pre">str</span></code>-<code class="docutils literal"><span class="pre">str</span></code> pairs</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_string_option</span><span class="p">(</span><span class="s2">&quot;hessian_approximation&quot;</span><span class="p">,</span><span class="s2">&quot;limited-memory&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_string_options</span><span class="p">()</span>
<span class="go">{&#39;hessian_approximation&#39;: &#39;limited-memory&#39;}</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.ipopt.replacement">
<code class="descname">replacement</code><a class="headerlink" href="#pygmo.ipopt.replacement" title="Permalink to this definition">¶</a></dt>
<dd><p>Individual replacement policy.</p>
<p>This attribute represents the policy that is used in the <code class="docutils literal"><span class="pre">evolve()</span></code> method to select the individual
that will be replaced by the optimised individual. The attribute can be either a string or an integral.</p>
<p>If the attribute is a string, it must be one of <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> will select the best individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> will select the worst individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> will randomly choose one individual in the population.</li>
</ul>
<p><a class="reference internal" href="#pygmo.ipopt.set_random_sr_seed" title="pygmo.ipopt.set_random_sr_seed"><code class="xref py py-func docutils literal"><span class="pre">set_random_sr_seed()</span></code></a> can be used to seed the random number generator
used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> policy.</p>
<p>If the attribute is an integer, it represents the index (in the population) of the individual that will be
replaced by the optimised individual.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the individual replacement policy or index</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">str</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if the attribute is set to an invalid string</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code></a> – if the attribute is set to a value of an invalid type</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.reset_integer_options">
<code class="descname">reset_integer_options</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.reset_integer_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear all integer options.</p>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_integer_option</span><span class="p">(</span><span class="s2">&quot;print_level&quot;</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_integer_options</span><span class="p">()</span>
<span class="go">{&#39;print_level&#39;: 3}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">reset_integer_options</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_integer_options</span><span class="p">()</span>
<span class="go">{}</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.reset_numeric_options">
<code class="descname">reset_numeric_options</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.reset_numeric_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear all numeric options.</p>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_numeric_option</span><span class="p">(</span><span class="s2">&quot;tol&quot;</span><span class="p">,</span><span class="mf">1E-4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_numeric_options</span><span class="p">()</span> 
<span class="go">{&#39;tol&#39;: 1E-4}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">reset_numeric_options</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_numeric_options</span><span class="p">()</span>
<span class="go">{}</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.reset_string_options">
<code class="descname">reset_string_options</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.reset_string_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Clear all string options.</p>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_string_option</span><span class="p">(</span><span class="s2">&quot;hessian_approximation&quot;</span><span class="p">,</span><span class="s2">&quot;limited-memory&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_string_options</span><span class="p">()</span>
<span class="go">{&#39;hessian_approximation&#39;: &#39;limited-memory&#39;}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">reset_string_options</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">get_string_options</span><span class="p">()</span>
<span class="go">{}</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="pygmo.ipopt.selection">
<code class="descname">selection</code><a class="headerlink" href="#pygmo.ipopt.selection" title="Permalink to this definition">¶</a></dt>
<dd><p>Individual selection policy.</p>
<p>This attribute represents the policy that is used in the <code class="docutils literal"><span class="pre">evolve()</span></code> method to select the individual
that will be optimised. The attribute can be either a string or an integral.</p>
<p>If the attribute is a string, it must be one of <code class="docutils literal"><span class="pre">&quot;best&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> and <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">&quot;best&quot;</span></code> will select the best individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;worst&quot;</span></code> will select the worst individual in the population,</li>
<li><code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> will randomly choose one individual in the population.</li>
</ul>
<p><a class="reference internal" href="#pygmo.ipopt.set_random_sr_seed" title="pygmo.ipopt.set_random_sr_seed"><code class="xref py py-func docutils literal"><span class="pre">set_random_sr_seed()</span></code></a> can be used to seed the random number generator
used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> policy.</p>
<p>If the attribute is an integer, it represents the index (in the population) of the individual that is selected
for optimisation.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">the individual selection policy or index</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first"><code class="docutils literal"><span class="pre">int</span></code> or <code class="docutils literal"><span class="pre">str</span></code></p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if the attribute is set to an invalid string</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#TypeError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">TypeError</span></code></a> – if the attribute is set to a value of an invalid type</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.set_integer_option">
<code class="descname">set_integer_option</code><span class="sig-paren">(</span><em>name</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.set_integer_option" title="Permalink to this definition">¶</a></dt>
<dd><p>Set integer option.</p>
<p>This method will set the optimisation integer option <em>name</em> to <em>value</em>.
The optimisation options are passed to the Ipopt API when calling the <code class="docutils literal"><span class="pre">evolve()</span></code> method.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>name</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – the name of the option</li>
<li><strong>value</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – the value of the option</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_integer_option</span><span class="p">(</span><span class="s2">&quot;print_level&quot;</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algorithm</span><span class="p">(</span><span class="n">ip</span><span class="p">)</span> 
<span class="go">Algorithm name: Ipopt: Interior Point Optimization [deterministic]</span>
<span class="go">        Thread safety: none</span>

<span class="go">Extra info:</span>
<span class="go">        Last optimisation return code: Solve_Succeeded (value = 0)</span>
<span class="go">        Verbosity: 0</span>
<span class="go">        Individual selection policy: best</span>
<span class="go">        Individual replacement policy: best</span>
<span class="go">        Integer options: {print_level : 3}</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.set_integer_options">
<code class="descname">set_integer_options</code><span class="sig-paren">(</span><em>opts</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.set_integer_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Set integer options.</p>
<p>This method will set the optimisation integer options contained in <em>opts</em>.
It is equivalent to calling <a class="reference internal" href="#pygmo.ipopt.set_integer_option" title="pygmo.ipopt.set_integer_option"><code class="xref py py-func docutils literal"><span class="pre">set_integer_option()</span></code></a> passing all the name-value pairs in <em>opts</em>
as arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>opts</strong> (<code class="docutils literal"><span class="pre">dict</span></code> of <code class="docutils literal"><span class="pre">str</span></code>-<code class="docutils literal"><span class="pre">int</span></code> pairs) – the name-value map that will be used to set the options</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_integer_options</span><span class="p">({</span><span class="s2">&quot;filter_reset_trigger&quot;</span><span class="p">:</span><span class="mi">4</span><span class="p">,</span> <span class="s2">&quot;print_level&quot;</span><span class="p">:</span><span class="mi">3</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algorithm</span><span class="p">(</span><span class="n">ip</span><span class="p">)</span> 
<span class="go">Algorithm name: Ipopt: Interior Point Optimization [deterministic]</span>
<span class="go">        Thread safety: none</span>

<span class="go">Extra info:</span>
<span class="go">        Last optimisation return code: Solve_Succeeded (value = 0)</span>
<span class="go">        Verbosity: 0</span>
<span class="go">        Individual selection policy: best</span>
<span class="go">        Individual replacement policy: best</span>
<span class="go">        Integer options: {filter_reset_trigger : 4,  print_level : 3}</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.set_numeric_option">
<code class="descname">set_numeric_option</code><span class="sig-paren">(</span><em>name</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.set_numeric_option" title="Permalink to this definition">¶</a></dt>
<dd><p>Set numeric option.</p>
<p>This method will set the optimisation numeric option <em>name</em> to <em>value</em>.
The optimisation options are passed to the Ipopt API when calling the <code class="docutils literal"><span class="pre">evolve()</span></code> method.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>name</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – the name of the option</li>
<li><strong>value</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – the value of the option</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_numeric_option</span><span class="p">(</span><span class="s2">&quot;tol&quot;</span><span class="p">,</span><span class="mf">1E-6</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algorithm</span><span class="p">(</span><span class="n">ip</span><span class="p">)</span> 
<span class="go">Algorithm name: Ipopt: Interior Point Optimization [deterministic]</span>
<span class="go">        Thread safety: none</span>

<span class="go">Extra info:</span>
<span class="go">        Last optimisation return code: Solve_Succeeded (value = 0)</span>
<span class="go">        Verbosity: 0</span>
<span class="go">        Individual selection policy: best</span>
<span class="go">        Individual replacement policy: best</span>
<span class="go">        Numeric options: {tol : 1E-6}</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.set_numeric_options">
<code class="descname">set_numeric_options</code><span class="sig-paren">(</span><em>opts</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.set_numeric_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Set numeric options.</p>
<p>This method will set the optimisation numeric options contained in <em>opts</em>.
It is equivalent to calling <a class="reference internal" href="#pygmo.ipopt.set_numeric_option" title="pygmo.ipopt.set_numeric_option"><code class="xref py py-func docutils literal"><span class="pre">set_numeric_option()</span></code></a> passing all the name-value pairs in <em>opts</em>
as arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>opts</strong> (<code class="docutils literal"><span class="pre">dict</span></code> of <code class="docutils literal"><span class="pre">str</span></code>-<code class="docutils literal"><span class="pre">float</span></code> pairs) – the name-value map that will be used to set the options</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_numeric_options</span><span class="p">({</span><span class="s2">&quot;tol&quot;</span><span class="p">:</span><span class="mf">1E-4</span><span class="p">,</span> <span class="s2">&quot;constr_viol_tol&quot;</span><span class="p">:</span><span class="mf">1E-3</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algorithm</span><span class="p">(</span><span class="n">ip</span><span class="p">)</span> 
<span class="go">Algorithm name: Ipopt: Interior Point Optimization [deterministic]</span>
<span class="go">        Thread safety: none</span>

<span class="go">Extra info:</span>
<span class="go">        Last optimisation return code: Solve_Succeeded (value = 0)</span>
<span class="go">        Verbosity: 0</span>
<span class="go">        Individual selection policy: best</span>
<span class="go">        Individual replacement policy: best</span>
<span class="go">        Numeric options: {constr_viol_tol : 1E-3,  tol : 1E-4}</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.set_random_sr_seed">
<code class="descname">set_random_sr_seed</code><span class="sig-paren">(</span><em>seed</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.set_random_sr_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the seed for the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code> selection/replacement policies.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – the value that will be used to seed the random number generator used by the <code class="docutils literal"><span class="pre">&quot;random&quot;</span></code>
election/replacement policies (see <a class="reference internal" href="#pygmo.ipopt.selection" title="pygmo.ipopt.selection"><code class="xref py py-attr docutils literal"><span class="pre">selection</span></code></a> and
<a class="reference internal" href="#pygmo.ipopt.replacement" title="pygmo.ipopt.replacement"><code class="xref py py-attr docutils literal"><span class="pre">replacement</span></code></a>)</p>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if the attribute is set to an integer which is negative or too large</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.set_string_option">
<code class="descname">set_string_option</code><span class="sig-paren">(</span><em>name</em>, <em>value</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.set_string_option" title="Permalink to this definition">¶</a></dt>
<dd><p>Set string option.</p>
<p>This method will set the optimisation string option <em>name</em> to <em>value</em>.
The optimisation options are passed to the Ipopt API when calling the <code class="docutils literal"><span class="pre">evolve()</span></code> method.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>name</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – the name of the option</li>
<li><strong>value</strong> (<code class="docutils literal"><span class="pre">str</span></code>) – the value of the option</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><p class="first last"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_string_option</span><span class="p">(</span><span class="s2">&quot;hessian_approximation&quot;</span><span class="p">,</span><span class="s2">&quot;limited-memory&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algorithm</span><span class="p">(</span><span class="n">ip</span><span class="p">)</span> 
<span class="go">Algorithm name: Ipopt: Interior Point Optimization [deterministic]</span>
<span class="go">    Thread safety: none</span>

<span class="go">Extra info:</span>
<span class="go">    Last optimisation return code: Solve_Succeeded (value = 0)</span>
<span class="go">    Verbosity: 0</span>
<span class="go">    Individual selection policy: best</span>
<span class="go">    Individual replacement policy: best</span>
<span class="go">    String options: {hessian_approximation : limited-memory}</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="pygmo.ipopt.set_string_options">
<code class="descname">set_string_options</code><span class="sig-paren">(</span><em>opts</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ipopt.set_string_options" title="Permalink to this definition">¶</a></dt>
<dd><p>Set string options.</p>
<p>This method will set the optimisation string options contained in <em>opts</em>.
It is equivalent to calling <a class="reference internal" href="#pygmo.ipopt.set_string_option" title="pygmo.ipopt.set_string_option"><code class="xref py py-func docutils literal"><span class="pre">set_string_option()</span></code></a> passing all the name-value pairs in <em>opts</em>
as arguments.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>opts</strong> (<code class="docutils literal"><span class="pre">dict</span></code> of <code class="docutils literal"><span class="pre">str</span></code>-<code class="docutils literal"><span class="pre">str</span></code> pairs) – the name-value map that will be used to set the options</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python (e.g.,
type conversion errors, mismatched function signatures, etc.)</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span> <span class="o">=</span> <span class="n">ipopt</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ip</span><span class="o">.</span><span class="n">set_string_options</span><span class="p">({</span><span class="s2">&quot;hessian_approximation&quot;</span><span class="p">:</span><span class="s2">&quot;limited-memory&quot;</span><span class="p">,</span> <span class="s2">&quot;limited_memory_initialization&quot;</span><span class="p">:</span><span class="s2">&quot;scalar1&quot;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algorithm</span><span class="p">(</span><span class="n">ip</span><span class="p">)</span> 
<span class="go">Algorithm name: Ipopt: Interior Point Optimization [deterministic]</span>
<span class="go">        Thread safety: none</span>

<span class="go">Extra info:</span>
<span class="go">        Last optimisation return code: Solve_Succeeded (value = 0)</span>
<span class="go">        Verbosity: 0</span>
<span class="go">        Individual selection policy: best</span>
<span class="go">        Individual replacement policy: best</span>
<span class="go">        String options: {hessian_approximation : limited-memory,  limited_memory_initialization : scalar1}</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.ihs">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">ihs</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>phmcr = 0.85</em>, <em>ppar_min = 0.35</em>, <em>ppar_max=0.99</em>, <em>bw_min=1e-5</em>, <em>bw_max=1.</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ihs" title="Permalink to this definition">¶</a></dt>
<dd><p>Harmony search (HS) is a metaheuristic algorithm said to mimick the improvisation process of musicians.
In the metaphor, each musician (i.e., each variable) plays (i.e., generates) a note (i.e., a value)
for finding a best harmony (i.e., the global optimum) all together.</p>
<p>This pygmo UDA implements the so-called improved harmony search algorithm (IHS), in which the probability
of picking the variables from the decision vector and the amount of mutation to which they are subject
vary (respectively linearly and exponentially) at each call of the <code class="docutils literal"><span class="pre">evolve()</span></code> method.</p>
<p>In this algorithm the number of fitness function evaluations is equal to the number of iterations.
All the individuals in the input population participate in the evolution. A new individual is generated
at every iteration, substituting the current worst individual of the population if better.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The HS algorithm can and has been  criticized, not for its performances,
but for the use of a metaphor that does not add anything to existing ones. The HS
algorithm essentially applies mutation and crossover operators to a background population and as such
should have been developed in the context of Evolutionary Strategies or Genetic Algorithms and studied
in that context. The use of the musicians metaphor only obscures its internal functioning
making theoretical results from ES and GA erroneously seem as unapplicable to HS.</p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The original IHS algorithm was designed to solve unconstrained, deterministic single objective problems.
In pygmo, the algorithm was modified to tackle also multi-objective, constrained (box and non linearly).
Such extension is original with pygmo.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations to consider (each generation will compute the objective function once)</li>
<li><strong>phmcr</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – probability of choosing from memory (similar to a crossover probability)</li>
<li><strong>ppar_min</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – minimum pitch adjustment rate. (similar to a mutation rate)</li>
<li><strong>ppar_max</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – maximum pitch adjustment rate. (similar to a mutation rate)</li>
<li><strong>bw_min</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – minimum distance bandwidth. (similar to a mutation width)</li>
<li><strong>bw_max</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – maximum distance bandwidth. (similar to a mutation width)</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em> or <em>seed</em> are negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>phmcr</em> is not in the ]0,1[ interval, <em>ppar_min</em> or <em>ppar_max</em> are not in the ]0,1[
interval, min/max quantities are less than/greater than max/min quantities, <em>bw_min</em> is negative.</li>
<li><code class="xref py py-exc docutils literal"><span class="pre">unspecified</span></code> – any exception thrown by failures at the intersection between C++ and Python
(e.g., type conversion errors, mismatched function signatures, etc.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/ihs.html#_CPPv2N5pagmo3ihsE" title="pagmo::ihs"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::ihs</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.ihs.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ihs.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code> and printed to screen.
The log frequency depends on the verbosity parameter (by default nothing is logged) which can be set calling
the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a> constructed with a
<a class="reference internal" href="#pygmo.ihs" title="pygmo.ihs"><code class="xref py py-class docutils literal"><span class="pre">ihs</span></code></a>.
A verbosity larger than 1 will produce a log with one entry each verbosity fitness evaluations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">ppar</span></code>, <code class="docutils literal"><span class="pre">bw</span></code>, <code class="docutils literal"><span class="pre">dx</span></code>, <code class="docutils literal"><span class="pre">df</span></code>,  <code class="docutils literal"><span class="pre">Violated</span></code>, <code class="docutils literal"><span class="pre">Viol.</span> <span class="pre">Norm</span></code>,``ideal``<ul class="simple">
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made.</li>
<li><code class="docutils literal"><span class="pre">ppar</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the pitch adjustment rate.</li>
<li><code class="docutils literal"><span class="pre">bw</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the distance bandwidth.</li>
<li><code class="docutils literal"><span class="pre">dx</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the population flatness evaluated as the distance between the decisions vector of the best and of the worst individual (or -1 in a multiobjective case).</li>
<li><code class="docutils literal"><span class="pre">df</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the population flatness evaluated as the distance between the fitness of the best and of the worst individual (or -1 in a multiobjective case).</li>
<li><code class="docutils literal"><span class="pre">Violated</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), the number of constraints violated by the current decision vector.</li>
<li><code class="docutils literal"><span class="pre">Viol.</span> <span class="pre">Norm</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the constraints violation norm for the current decision vector.</li>
<li><code class="docutils literal"><span class="pre">ideal_point</span></code> (1D numpy array), the ideal point of the current population (cropped to max 5 dimensions only in the screen output)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">ihs</span><span class="p">(</span><span class="mi">20000</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">2000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">hock_schittkowsky_71</span><span class="p">())</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span><span class="o">.</span><span class="n">c_tol</span> <span class="o">=</span> <span class="p">[</span><span class="mf">1e-1</span><span class="p">]</span><span class="o">*</span><span class="mi">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Fevals:          ppar:            bw:            dx:            df:      Violated:    Viol. Norm:        ideal1:</span>
<span class="go">      1       0.350032       0.999425        4.88642        14.0397              0              0        43.2982</span>
<span class="go">   2001       0.414032       0.316046        5.56101        25.7009              0              0        33.4251</span>
<span class="go">   4001       0.478032      0.0999425          5.036        26.9657              0              0        19.0052</span>
<span class="go">   6001       0.542032      0.0316046        3.77292        23.9992              0              0        19.0052</span>
<span class="go">   8001       0.606032     0.00999425        3.97937        16.0803              0              0        18.1803</span>
<span class="go">  10001       0.670032     0.00316046        1.15023        1.57947              0              0        17.8626</span>
<span class="go">  12001       0.734032    0.000999425       0.017882      0.0185438              0              0        17.5894</span>
<span class="go">  14001       0.798032    0.000316046     0.00531358      0.0074745              0              0        17.5795</span>
<span class="go">  16001       0.862032    9.99425e-05     0.00270865     0.00155563              0              0        17.5766</span>
<span class="go">  18001       0.926032    3.16046e-05     0.00186637     0.00167523              0              0        17.5748</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">ihs</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 0.35003234534534, 0.9994245193792801, 4.886415773459253, 14.0397487316794, ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/ihs.html#_CPPv2NK5pagmo3ihs7get_logEv" title="pagmo::ihs::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::ihs::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.ihs.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.ihs.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

<hr class="docutils" />
<dl class="class">
<dt id="pygmo.xnes">
<em class="property">class </em><code class="descclassname">pygmo.</code><code class="descname">xnes</code><span class="sig-paren">(</span><em>gen = 1</em>, <em>eta_mu = -1</em>, <em>eta_sigma = -1</em>, <em>eta_b = -1</em>, <em>sigma0 = -1</em>, <em>ftol = 1e-6</em>, <em>xtol = 1e-6</em>, <em>memory = False</em>, <em>force_bounds = False</em>, <em>seed = random</em><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.xnes" title="Permalink to this definition">¶</a></dt>
<dd><p>Exponential Evolution Strategies.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>gen</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – number of generations</li>
<li><strong>eta_mu</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – learning rate for mean update (if -1 will be automatically selected to be 1)</li>
<li><strong>eta_sigma</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – learning rate for step-size update (if -1 will be automatically selected)</li>
<li><strong>eta_b</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – learning rate for the covariance matrix update (if -1 will be automatically selected)</li>
<li><strong>sigma0</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – the initial search width will be sigma0 * (ub - lb) (if -1 will be automatically selected to be 1)</li>
<li><strong>ftol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the x tolerance</li>
<li><strong>xtol</strong> (<code class="docutils literal"><span class="pre">float</span></code>) – stopping criteria on the f tolerance</li>
<li><strong>memory</strong> (<code class="docutils literal"><span class="pre">bool</span></code>) – when true the adapted parameters are not reset between successive calls to the evolve method</li>
<li><strong>force_bounds</strong> (<code class="docutils literal"><span class="pre">bool</span></code>) – when true the box bounds are enforced. The fitness will never be called outside the bounds but the covariance matrix adaptation  mechanism will worsen</li>
<li><strong>seed</strong> (<code class="docutils literal"><span class="pre">int</span></code>) – seed used by the internal random number generator (default is random)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Raises:</th><td class="field-body"><ul class="first last simple">
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#OverflowError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">OverflowError</span></code></a> – if <em>gen</em> is negative or greater than an implementation-defined value</li>
<li><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.7)"><code class="xref py py-exc docutils literal"><span class="pre">ValueError</span></code></a> – if <em>eta_mu</em>, <em>eta_sigma</em>, <em>eta_b</em>, <em>sigma0</em> are not in ]0,1] or -1</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p>See also the docs of the C++ class <a class="reference internal" href="../../cpp/algorithms/xnes.html#_CPPv2N5pagmo4xnesE" title="pagmo::xnes"><code class="xref cpp cpp-class docutils literal"><span class="pre">pagmo::xnes</span></code></a>.</p>
<dl class="method">
<dt id="pygmo.xnes.get_log">
<code class="descname">get_log</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.xnes.get_log" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a log containing relevant parameters recorded during the last call to <code class="docutils literal"><span class="pre">evolve()</span></code>. The log frequency depends on the verbosity
parameter (by default nothing is logged) which can be set calling the method <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm.set_verbosity" title="pygmo.algorithm.set_verbosity"><code class="xref py py-func docutils literal"><span class="pre">set_verbosity()</span></code></a> on an <a class="reference internal" href="../py_algorithm.html#pygmo.algorithm" title="pygmo.algorithm"><code class="xref py py-class docutils literal"><span class="pre">algorithm</span></code></a>
constructed with a <a class="reference internal" href="#pygmo.xnes" title="pygmo.xnes"><code class="xref py py-class docutils literal"><span class="pre">xnes</span></code></a>. A verbosity of <code class="docutils literal"><span class="pre">N</span></code> implies a log line each <code class="docutils literal"><span class="pre">N</span></code> generations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">at each logged epoch, the values <code class="docutils literal"><span class="pre">Gen</span></code>, <code class="docutils literal"><span class="pre">Fevals</span></code>, <code class="docutils literal"><span class="pre">Best</span></code>, <code class="docutils literal"><span class="pre">dx</span></code>, <code class="docutils literal"><span class="pre">df</span></code>, <code class="docutils literal"><span class="pre">sigma</span></code>, where:<ul class="simple">
<li><code class="docutils literal"><span class="pre">Gen</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), generation number</li>
<li><code class="docutils literal"><span class="pre">Fevals</span></code> (<code class="docutils literal"><span class="pre">int</span></code>), number of functions evaluation made</li>
<li><code class="docutils literal"><span class="pre">Best</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the best fitness function currently in the population</li>
<li><code class="docutils literal"><span class="pre">dx</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the norm of the distance to the population mean of the mutant vectors</li>
<li><code class="docutils literal"><span class="pre">df</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the population flatness evaluated as the distance between the fitness of the best and of the worst individual</li>
<li><code class="docutils literal"><span class="pre">sigma</span></code> (<code class="docutils literal"><span class="pre">float</span></code>), the current step-size</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">list</span></code> of <code class="docutils literal"><span class="pre">tuples</span></code></td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">pygmo</span> <span class="k">import</span> <span class="o">*</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span> <span class="o">=</span> <span class="n">algorithm</span><span class="p">(</span><span class="n">xnes</span><span class="p">(</span><span class="n">gen</span> <span class="o">=</span> <span class="mi">500</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">algo</span><span class="o">.</span><span class="n">set_verbosity</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">prob</span> <span class="o">=</span> <span class="n">problem</span><span class="p">(</span><span class="n">rosenbrock</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">population</span><span class="p">(</span><span class="n">prob</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pop</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">evolve</span><span class="p">(</span><span class="n">pop</span><span class="p">)</span> 
<span class="go">Gen:        Fevals:          Best:            dx:            df:         sigma:</span>
<span class="go">  1              0         173924        33.6872    3.06519e+06            0.5</span>
<span class="go">101           2000        92.9612       0.583942        156.921      0.0382078</span>
<span class="go">201           4000        8.79819       0.117574          5.101      0.0228353</span>
<span class="go">301           6000        4.81377      0.0698366        1.34637      0.0297664</span>
<span class="go">401           8000        1.04445      0.0568541       0.514459      0.0649836</span>
<span class="go">Exit condition -- generations = 500</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span> <span class="o">=</span> <span class="n">algo</span><span class="o">.</span><span class="n">extract</span><span class="p">(</span><span class="n">xnes</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">uda</span><span class="o">.</span><span class="n">get_log</span><span class="p">()</span> 
<span class="go">[(1, 0, 173924.2840042722, 33.68717961390855, 3065192.3843070837, 0.5), ...</span>
</pre></div>
</div>
<p>See also the docs of the relevant C++ method <a class="reference internal" href="../../cpp/algorithms/xnes.html#_CPPv2NK5pagmo4xnes7get_logEv" title="pagmo::xnes::get_log"><code class="xref cpp cpp-func docutils literal"><span class="pre">pagmo::xnes::get_log()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="pygmo.xnes.get_seed">
<code class="descname">get_seed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pygmo.xnes.get_seed" title="Permalink to this definition">¶</a></dt>
<dd><p>This method will return the random seed used internally by this uda.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the random seed of the population</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><code class="docutils literal"><span class="pre">int</span></code></td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2017-2018, pagmo development team.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.7.<br/>
    </p>
  </div>
</footer>
  </body>
</html>